// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/consts.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DUST_LIMIT = exports.TX_OUTPUT_SATOSHI_BYTE_LEN = exports.TX_INPUT_PREV_TX_HASH_BYTE_LEN = exports.TX_VERSION_BYTE_LEN = exports.TX_INPUT_SCRIPT_HASH_BYTE_LEN = exports.TX_OUTPUT_BYTE_LEN = exports.TX_INPUT_BYTE_LEN = exports.TX_OUTPUT_DATA_HASH_LEN = exports.TX_OUTPUT_SCRIPT_HASH_LEN = exports.HASH160_HASH_LEN = exports.RIPEMD160_HASH_LEN = exports.TX_HASH_BYTE_LEN = exports.HASH256_HASH_LEN = exports.SHA256_HASH_LEN = void 0;
/**
 * @ignore
 */
exports.SHA256_HASH_LEN = 32n;
/** @ignore */
exports.HASH256_HASH_LEN = exports.SHA256_HASH_LEN;
/** @ignore */
exports.TX_HASH_BYTE_LEN = exports.HASH256_HASH_LEN;
/** @ignore */
exports.RIPEMD160_HASH_LEN = 20n;
/** @ignore */
exports.HASH160_HASH_LEN = exports.RIPEMD160_HASH_LEN;
/** @ignore */
exports.TX_OUTPUT_SCRIPT_HASH_LEN = exports.SHA256_HASH_LEN;
/** @ignore */
exports.TX_OUTPUT_DATA_HASH_LEN = exports.SHA256_HASH_LEN;
/** @ignore
 * prevout (32 + 4) + unlockScriptHash(32) + sequence(4) = 72
 */
exports.TX_INPUT_BYTE_LEN = 72n;
/** @ignore
 * satoshis(8) + scriptHash(32) + dataHash(32) = 72
 */
exports.TX_OUTPUT_BYTE_LEN = 72n;
/** @ignore */
exports.TX_INPUT_SCRIPT_HASH_BYTE_LEN = 32n;
/** @ignore byte length of each part in tx */
exports.TX_VERSION_BYTE_LEN = 4n;
/** @ignore */
exports.TX_INPUT_PREV_TX_HASH_BYTE_LEN = exports.TX_HASH_BYTE_LEN;
/** @ignore */
exports.TX_OUTPUT_SATOSHI_BYTE_LEN = 8n;
/** @ignore the dust limit is the minimum amount of satoshis that can be sent in a transaction */
exports.DUST_LIMIT = 1;
