// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/smartContract.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmartContract = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
const abstractContract_js_1 = require('./abstractContract.cjs');
const buildOutput_js_1 = require('./methods/buildOutput.cjs');
const checkCtx_js_1 = require('./methods/checkCtx.cjs');
const checkSHPreimage_js_1 = require('./methods/checkSHPreimage.cjs');
const checkSig_js_1 = require('./methods/checkSig.cjs');
const checkDataSig_js_1 = require('./methods/checkDataSig.cjs');
const abi_js_1 = require('./abi.cjs');
const script_js_1 = require('./types/script.cjs');
const index_js_1 = require('../utils/index.cjs');
const hashes_js_1 = require('./fns/hashes.cjs');
const byteString_js_1 = require('./fns/byteString.cjs');
const checkInputState_js_1 = require('./methods/checkInputState.cjs');
const serializer_js_1 = require('./serializer.cjs');
const backtraceToGenensis_js_1 = require('./methods/backtraceToGenensis.cjs');
const stateSerializer_js_1 = require('./stateSerializer.cjs');
const abiutils_js_1 = require('./abiutils.cjs');
const checkInputStateHashes_js_1 = require('./methods/checkInputStateHashes.cjs');
const proof_js_1 = require('../utils/proof.cjs');
const sigUtils_js_1 = require('../utils/sigUtils.cjs');
const assert_js_1 = require('./fns/assert.cjs');
/**
 * The main contract class. To write a contract, extend this class as such:
 * @onchain
 * @example
 *  ```ts
 * class YourSmartContract extends SmartContract {
 *   // your smart contract code here
 * }
 * ```
 * @category SmartContract
 */
class SmartContract extends abstractContract_js_1.AbstractContract {
    /**
     *
     * @ignore
     */
    getStateType() {
        const clazz = this.constructor;
        return clazz.stateType;
    }
    /**
     * Get the data of the contract.
     * @returns the data of the contract
     */
    get data() {
        return this._data;
    }
    /**
     * Set the data of the contract.
     * Only stateless contracts (without stateType) can set data.
     * Stateful contracts should use the state property instead.
     * @param value the data to set
     */
    set data(value) {
        if (this.getStateType() !== undefined) {
            throw new Error(`Cannot set data on stateful contract '${this.constructor.name}'. ` +
                `Use 'state' property instead for stateful contracts.`);
        }
        this._data = value;
    }
    /**
     * This function is usually called on the frontend.
     * The contract class needs to call this function before instantiating.
     * @param artifact a contract artifact json object
     */
    static loadArtifact(artifact) {
        if ((0, abiutils_js_1.getUnRenamedSymbol)(artifact.contract) !== (0, abiutils_js_1.getUnRenamedSymbol)(this.name)) {
            throw new Error(`Artifact does not match contract ${this.name}!`);
        }
        this.artifact = artifact;
        this.stateType = artifact.stateType;
        return this;
    }
    static get artifactHexMD5() {
        return (0, index_js_1.calcArtifactHexMD5)(this.artifact.hex);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    constructor(...args) {
        super();
        const ContractClazz = this.constructor;
        if (!ContractClazz.artifact) {
            throw new Error(`Artifact not loaded for contract ${this.constructor.name}!`);
        }
        const artifact = ContractClazz.artifact;
        this._abiCoder = new abi_js_1.ABICoder(artifact);
        if (Object.getPrototypeOf(this.constructor.prototype) !== SmartContract.prototype) {
            // contract is not a direct child of SmartContract, should use ContractClazz.create instead
            if (!SmartContract.newFromCreate) {
                throw new Error(`Contract \`${this.constructor.name}\` is not a direct child of \`SmartContract\`, should use \`${this.constructor.name}.create\` instead of use \`new ${this.constructor.name}\``);
            }
        }
        if (!SmartContract.newFromCreate) {
            this.initLockingScript(...args);
        }
        // init
        this._data = (0, byteString_js_1.toByteString)('');
    }
    /**
   * @ignore
   * @param args
   */
    initLockingScript(...args) {
        this._lockingScriptHexWithoutHeader = this._abiCoder.encodeConstructorCall(args).toHex();
        // default: add md5, do not add tag
        this.contractHeader = {
            version: index_js_1.ContractHeaderSerializer.VERSION,
            tags: this.constructor.tags || [],
            md5: this.constructor.artifactHexMD5,
        };
        this.updateLockingScript();
    }
    updateLockingScript() {
        if (!this.contractHeader) {
            this.lockingScript = script_js_1.Script.fromHex(this._lockingScriptHexWithoutHeader);
            return;
        }
        const txOutScriptHex = index_js_1.ContractHeaderSerializer.serialize(this.contractHeader, this._lockingScriptHexWithoutHeader);
        const txOutScript = script_js_1.Script.fromHex(txOutScriptHex);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        this.lockingScript = txOutScript;
    }
    /**
     * Used to create an instance of a smart contract. If your smart contract has a base class, you must use this method to instantiate it.
     * @param this
     * @param args
     * @returns
     */
    static create(...args) {
        SmartContract.newFromCreate = true;
        const instance = new this(...args);
        SmartContract.newFromCreate = false;
        instance.initLockingScript(...args);
        return instance;
    }
    /**
     * Using the [OP_PUSH_TX]{@link https://medium.com/@xiaohuiliu/op-push-tx-3d3d279174c1} technique, check if `shPreimage` is the preimage of the current transaction.
     * @onchain
     * @param txPreimage The format of the preimage is [specified]{@link https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#signature-validation-rules}
     * @returns true if `shPreimage` is the preimage of the current transaction. Otherwise false.
     */
    checkSHPreimage(shPreimage) {
        return (0, checkSHPreimage_js_1.checkSHPreimageImpl)(this, shPreimage);
    }
    /**
     * A built-in function verifies an Schnorr signature. It takes two inputs from the stack, a public key (on top of the stack) and an Schnorr signature in its DER_CANONISED format concatenated with sighash flags.
     * It outputs true or false on the stack based on whether the signature check passes or fails. [see]{@link https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#specification}
     * @onchain
     * @category Signature Verification
     */
    checkSig(signature, publickey, errorMsg = 'signature check failed') {
        const fSuccess = (0, checkSig_js_1.checkSigImpl)(this, signature, publickey);
        if (!fSuccess && signature.length) {
            // because NULLFAIL rule, always throw if catch a wrong signature
            // https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki#nullfail
            throw new Error(errorMsg);
        }
        return fSuccess;
    }
    /**
     * Verifies an ECDSA signature with an explicit sighash flag check.
     * This method validates that the signature's trailing sighash flag byte matches
     * the explicitly provided flag parameter before performing the signature verification.
     *
     * @param signature - The signature to verify (DER encoded with sighash flag appended)
     * @param publickey - The public key to verify the signature against
     * @param sigHashFlag - The expected sighash flag value (e.g., 1n for ALL, 2n for NONE, 129n for ANYONECANPAY_ALL)
     * @returns true if the flag matches AND the signature is valid, false otherwise
     * @onchain
     * @category Signature Verification
     */
    checkSigWithFlag(signature, publickey, sigHashFlag) {
        // Extract the last byte of the signature (sighash flag)
        const sigHex = signature;
        const lastByte = parseInt(sigHex.slice(-2), 16);
        // Check if the flag matches
        if (BigInt(lastByte) !== sigHashFlag) {
            return false;
        }
        // Perform the actual signature check
        return (0, checkSig_js_1.checkSigImpl)(this, signature, publickey);
    }
    /**
     * Compares the first signature against each public key until it finds an ECDSA match.
     * Starting with the subsequent public key, it compares the second signature against
     * each remaining public key until it finds an ECDSA match. The process is repeated
     * until all signatures have been checked or not enough public keys remain
     * to produce a successful result. All signatures need to match a public key.
     * Because public keys are not checked again if they fail any signature comparison,
     * signatures must be placed in the scriptSig using the same order as their corresponding
     * public keys were placed in the scriptPubKey or redeemScript. If all signatures are
     * valid, 1 is returned, 0 otherwise. Due to a bug, one extra unused value is removed from the stack.
     * @onchain
     * @category Signature Verification
     */
    checkMultiSig(signatures, publickeys) {
        return (0, checkSig_js_1.checkMultiSigImpl)(this, signatures, publickeys);
    }
    /**
     * Verifies an ECDSA signature against an explicit message and public key.
     * Unlike checkSig which uses the transaction preimage as the implicit message,
     * checkDataSig allows verifying signatures on arbitrary data.
     *
     * Uses OP_CHECKSIGFROMSTACK (0xba) under the hood.
     * Stack order: <sig> <msg> <pubKey> (bottom to top)
     *
     * @param signature - The signature to verify (DER encoded with sighash type)
     * @param message - The message that was signed (will be SHA256 hashed once)
     * @param publickey - The public key to verify the signature against
     * @returns true if the signature is valid, false otherwise
     * @onchain
     * @category Signature Verification
     */
    checkDataSig(signature, message, publickey) {
        return (0, checkDataSig_js_1.checkDataSigImpl)(this, signature, message, publickey);
    }
    /**
     * A built-in function to create an [change output]{@link https://en.bitcoin.it/wiki/Change}.
     * @onchain
     * @returns
     */
    buildChangeOutput() {
        this._checkPsbtBinding();
        return (0, buildOutput_js_1.buildChangeOutputImpl)(this._curPsbt);
    }
    /**
     * A built-in function to create all outputs  added by `appendStateOutput()`
     * @onchain
     * @returns an output containing the new state
     */
    // buildStateOutput(satoshis: Int32): ByteString {
    //   this._checkPsbtBinding();
    //   return buildStateOutputImpl(this, this.state, satoshis, this.ctx.spentScriptHash);
    // }
    /**
     * Serializes the contract state into a ByteString.
     * @template T - Type of the contract state extending OpcatState
     * @param this - Constructor reference for type inference
     * @param state - The contract state object to serialize
     * @returns Serialized state as ByteString
     * @throws Error if artifact is not loaded or state type is not defined
     */
    static serializeState(state) {
        const selfClazz = this;
        const artifact = selfClazz.artifact;
        if (!artifact) {
            throw new Error(`Artifact is not loaded for the contract: ${this.name}`);
        }
        const stateType = selfClazz.stateType;
        if (!stateType) {
            throw new Error(`State struct \`${stateType}\` is not defined!`);
        }
        return (0, stateSerializer_js_1.serializeState)(artifact, stateType, state);
    }
    /**
     * Deserializes the contract state from a ByteString.
     * @template T - The type of the contract state extending OpcatState
     * @param serializedState - The serialized state as a ByteString
     * @returns The deserialized state object of type T
     * @throws Error if artifact is not loaded or state type is not defined
     */
    static deserializeState(serializedState) {
        const selfClazz = this;
        const artifact = selfClazz.artifact;
        if (!artifact) {
            throw new Error(`Artifact is not loaded for the contract: ${this.name}`);
        }
        const stateType = selfClazz.stateType;
        if (!stateType) {
            throw new Error(`State struct \`${stateType}\` is not defined!`);
        }
        return (0, stateSerializer_js_1.deserializeState)(artifact, stateType, serializedState);
    }
    /**
     * Computes the SHA-256 hash of the serialized contract state.
     *
     * @template T - Type extending OpcatState
     * @param state - The contract state to hash
     * @returns ByteString containing the SHA-256 hash of the serialized state
     * @override
     */
    static stateHash(state) {
        return (0, hashes_js_1.sha256)(this.serializeState(state));
    }
    /**
     * check state of the input.
     * option in the `@method()` decorator to false.
     * @onchain
     * @param inputIndex index of the input
     * @param serializedState the state of the input
     * @returns success if stateHash is valid
     */
    checkInputState(inputIndex, serializedState) {
        const _stateHash = (0, byteString_js_1.slice)(this.inputContext.spentDataHashes, inputIndex * 32n, (inputIndex + 1n) * 32n);
        (0, assert_js_1.assert)((0, hashes_js_1.sha256)(serializedState) == _stateHash, 'stateHash mismatch');
        return true;
    }
    /**
     * Get the context details of the current PSBT input in which this contract is called or spent.
     * @onchain
     */
    get ctx() {
        const { shPreimage, inputCount, prevouts, prevout, spentScriptHashes, spentAmounts, spentDataHashes, } = this.inputContext;
        return {
            ...shPreimage,
            inputCount,
            prevouts,
            prevout,
            spentScriptHashes,
            spentAmounts,
            spentDataHashes,
        };
    }
    /**
     * Mark the contract instance as spent from the input of the PSBT.
     * @param psbt
     * @param inputIndex
     */
    spentFromInput(psbt, inputIndex) {
        this._curPsbt = psbt;
        this._curInputIndex = inputIndex;
    }
    /**
     * Gets the PSBT (Partially Signed Bitcoin Transaction) that was used to spend this contract.
     * Returns undefined if no spending transaction exists.
     */
    get spentPsbt() {
        return this._curPsbt;
    }
    /**
     * Sets the signature hash type for the current input in the PSBT.
     * @param sigHashType - The signature hash type to set
     */
    setSighashType(sigHashType) {
        this._checkPsbtBinding();
        this._curPsbt.setSighashType(this._curInputIndex, sigHashType);
    }
    /**
     * Get the context of the current PSBT input in which this contract is called or spent.
     * @returns the context of the current PSBT input
     */
    get inputContext() {
        this._checkPsbtBinding();
        return this._curPsbt.getInputCtx(this._curInputIndex);
    }
    /**
     * Extends method arguments with context if needed and encodes them into a unlocking script.
     * @ignore
     * @param method - The method name to call.
     * @param args - The arguments to pass to the method.
     * @param autoCheckInputState - Whether to automatically check input state before injection.
     */
    extendMethodArgs(method, args, autoCheckInputState) {
        // extend the args with the context
        if (this._shouldInjectCtx(method)) {
            this._autoInject(method, args, autoCheckInputState);
        }
        const unlockingScript = this._abiCoder.encodePubFunctionCall(method, args);
        this._methodCall = {
            method,
            args,
            unlockingScript,
        };
    }
    /**
     * Checks if the given method is a public function in the smart contract.
     * @param method - The method name to check.
     * @returns True if the method is a public function, false otherwise.
     */
    isPubFunction(method) {
        return this._abiCoder.isPubFunction(method);
    }
    /**
     * @ignore
     * Checks if the current instance is a smart contract.
     * @returns {boolean} Always returns true indicating this is a smart contract.
     */
    isSmartContract() {
        return true;
    }
    _shouldInjectCtx(method) {
        return this._abiCoder.artifact.abi.some((abiEntity) => {
            return (abiEntity.name === method &&
                abiEntity.params.some((param) => param.name === '__scrypt_ts_shPreimage'));
        });
    }
    _autoInject(method, args, autoCheckInputState) {
        const { shPreimage, prevouts, prevout, spentScriptHashes, spentAmounts, spentDataHashes, inputCount, } = this.inputContext;
        // Generate and store preimage signature BEFORE checkCtxImpl
        // because checkCtxImpl calls checkSHPreimage which needs the signature.
        //
        // The signature is stored WITH sighash flag appended because:
        // - checkSig (OP_CHECKSIG) requires signature with sighash flag
        // - checkDataSig (OP_CHECKSIGFROMSTACK) requires pure DER signature without sighash flag
        //   (the sighash flag is stripped via slice in checkSHPreimageImpl and transpiled code)
        //
        // Extract the actual sighash type from shPreimage to support all sighash types
        // (SIGHASH_ALL, SIGHASH_NONE, SIGHASH_SINGLE, with or without ANYONECANPAY)
        const sigHashType = Number(shPreimage.sigHashType);
        this._injectedPreimageSig = (0, sigUtils_js_1.signSHPreimage)(shPreimage, sigHashType);
        (0, checkCtx_js_1.checkCtxImpl)(this, shPreimage, this._curInputIndex, prevouts, prevout, spentScriptHashes, spentAmounts, spentDataHashes);
        const curState = this.state;
        const abiEntity = this._abiCoder.artifact.abi.find((abiEntity) => {
            return abiEntity.name === method;
        });
        if (abiEntity) {
            abiEntity.params.forEach((param) => {
                if (param.name === '__scrypt_ts_shPreimage') {
                    const { shPreimage } = this.inputContext;
                    args.push(shPreimage);
                }
                else if (param.name === '__scrypt_ts_changeInfo') {
                    args.push(this.changeInfo);
                }
                else if (param.name === '__scrypt_ts_prevouts') {
                    args.push(prevouts);
                }
                else if (param.name === '__scrypt_ts_prevout') {
                    args.push(prevout);
                }
                else if (param.name === '__scrypt_ts_spentScriptHashes') {
                    const { spentScriptHashes } = this.inputContext;
                    args.push(spentScriptHashes);
                }
                else if (param.name === '__scrypt_ts_spentAmounts') {
                    args.push(spentAmounts);
                }
                else if (param.name === '__scrypt_ts_inputCount') {
                    args.push(inputCount);
                }
                else if (param.name === '__scrypt_ts_spentDataHashes') {
                    (0, checkInputStateHashes_js_1.checkInputStateHashesImpl)(shPreimage.hashSpentDataHashes, spentDataHashes);
                    args.push(spentDataHashes);
                }
                else if (param.name === '__scrypt_ts_curState') {
                    this._checkState();
                    if (autoCheckInputState) {
                        (0, checkInputState_js_1.checkInputStateImpl)(shPreimage.spentDataHash, this.constructor.serializeState(curState));
                    }
                    args.push(curState);
                }
                else if (param.name === '__scrypt_ts_prevTxHashPreimage') {
                    if (!this.utxo.txHashPreimage) {
                        throw new Error('utxo.txHashPreimage is required for backtrace');
                    }
                    args.push((0, proof_js_1.toTxHashPreimage)((0, index_js_1.hexToUint8Array)(this.utxo.txHashPreimage)));
                }
                else if (param.name === '__scrypt_ts_preimageSig') {
                    // Use the preimage signature that was already generated at the start of _autoInject
                    args.push(this._injectedPreimageSig);
                }
            });
        }
    }
    _checkPsbtBinding() {
        if (this._curPsbt === undefined || this._curInputIndex === undefined) {
            throw new Error('Not binding to any PSBT input!');
        }
    }
    _checkState() {
        if (this.state === undefined) {
            throw new Error('State is not initialized!');
        }
    }
    /**
     * Get the change info of the change output for current psbt.
     * @onchain
     * @returns the change info of current psbt
     */
    get changeInfo() {
        this._checkPsbtBinding();
        return this._curPsbt.getChangeInfo();
    }
    /**
     * A set of functions for debugging contracts, which can only be called in `@method` methods.
     * @onchain
     */
    get debug() {
        return {
            diffOutputs: (outputsByte) => {
                const outputs = (0, serializer_js_1.deserializeOutputs)(outputsByte);
                if (outputs.length !== this._curPsbt?.txOutputs.length) {
                    console.warn(`Outputs count is different: ${outputs.length}(#contract) vs ${this._curPsbt?.txOutputs.length}(#context)`);
                    return;
                }
                this._curPsbt?.txOutputs.forEach((output, index) => {
                    const scriptHash = (0, hashes_js_1.sha256)((0, index_js_1.uint8ArrayToHex)(output.script));
                    const dataHash = (0, hashes_js_1.sha256)((0, index_js_1.uint8ArrayToHex)(output.data));
                    if (outputs[index].dataHash !== dataHash) {
                        console.warn(`Output[${index}] dataHash is different: ${outputs[index]?.dataHash}(#contract) vs ${dataHash}(#context)`);
                    }
                    if (outputs[index].scriptHash !== scriptHash) {
                        console.warn(`Output[${index}] scriptHash is different: ${outputs[index]?.scriptHash}(#contract) vs ${scriptHash}(#context)`);
                    }
                    if (outputs[index]?.value !== output.value) {
                        console.warn(`Output[${index}] value is different: ${outputs[index]?.value}(#contract) vs ${output.value}(#context)`);
                    }
                });
            },
        };
    }
    /**
     * Check the outputs with the context of current transaction.
     * @param outputs the expected serialized outputs of the current transaction
     * @returns true if the outputs are not consistent with the transaction context, otherwise false.
     * @onchain
     */
    checkOutputs(outputs) {
        if (this.ctx.hashOutputs !== (0, hashes_js_1.hash256)(outputs)) {
            this.debug.diffOutputs(outputs);
            return false;
        }
        return true;
    }
    /**
     * Returns the raw arguments from the call data of the smart contract.
     * @returns The raw arguments extracted from the call data.
     */
    getUnlockingScript() {
        return this.getCallData().unlockingScript;
    }
    /**
     * Gets the method call data for the current smart contract.
     * @throws {Error} If no method call is found.
     * @returns {MethodCallData} The method call data object.
     */
    getCallData() {
        if (!this._methodCall) {
            throw new Error('No method call found!');
        }
        return this._methodCall;
    }
    /**
     * Transform unlocking script from the testimony into arguments previously used to invoke the contract.
     * @ignore
     * @param _args
     * @param _method
     */
    unlockingScriptToCallData(_args, _method) {
        throw new Error('Method not implemented.');
    }
    /**
     * Checks a locktime parameter with the transaction's locktime.
     * There are two times of nLockTime: lock-by-blockheight and lock-by-blocktime,
     * distinguished by whether nLockTime < LOCKTIME_THRESHOLD = 500000000
     *
     * See the corresponding code on bitcoin core:
     * https://github.com/bitcoin/bitcoin/blob/ffd75adce01a78b3461b3ff05bcc2b530a9ce994/src/script/interpreter.cpp#L1129
     *
     * See the bip65 for specification
     * https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki#summary
     * @onchain
     * @param {BN} nLockTime the locktime read from the script
     * @return {boolean} true if the transaction's locktime is less than or equal to
     *                   the transaction's locktime
     */
    timeLock(nLockTime) {
        // We want to compare apples to apples, so fail the script
        // unless the type of nLockTime being tested is the same as
        // the nLockTime in the transaction.
        const sequence = this._curPsbt.getSequence(this._curInputIndex);
        if (!(0, index_js_1.isFinal)(sequence)) {
            return false;
        }
        let res = true;
        // Check if using block height.
        if (nLockTime < 500000000) {
            // Enforce nLocktime field to also use block height.
            res = res && this.ctx.nLockTime < 500000000;
        }
        //return true;
        return res && this.ctx.nLockTime >= nLockTime;
    }
    /**
     * Check whether the contract can be traced back to the genesis outpoint.
     * @param backtraceInfo the backtrace info to verify, including prevTx and prevPrevTx informations
     * @param genesisOutpoint expected genesis outpoint to be traced back to
     * @returns true if the contract can be backtraced to the genesis outpoint. Otherwise false.
     * @onchain
     * @category Backtrace
     */
    backtraceToOutpoint(backtraceInfo, genesisOutpoint) {
        return (0, backtraceToGenensis_js_1.backtraceToOutpointImpl)(this, backtraceInfo, genesisOutpoint);
    }
    /**
     * Check whether the contract can be traced back to the genesis script.
     * @param backtraceInfo the backtrace info to verify, including prevTx and prevPrevTx informations
     * @param genesisScript expected genesis script to be traced back to
     * @returns true if the contract can be backtraced to the genesis script. Otherwise false.
     * @onchain
     * @category Backtrace
     */
    backtraceToScript(backtraceInfo, genesisScript) {
        return (0, backtraceToGenensis_js_1.backtraceToScriptImpl)(this, backtraceInfo, genesisScript);
    }
    /**
     * Get a new contract instance with the new state.
     * @param newState the new state
     * @returns the new covenant
     */
    next(newState) {
        const next = (0, index_js_1.cloneDeep)(this);
        next.state = newState;
        next.utxo = undefined;
        return next;
    }
    /**
     * Binds the smart contract to a UTXO by verifying and setting its script.
     * @param utxo - The UTXO to bind to (script field is optional)
     * @returns The contract instance for chaining
     * @throws Error if the UTXO's script exists and doesn't match the contract's locking script
     */
    bindToUtxo(utxo) {
        if (utxo.script && utxo.script !== this.lockingScript.toHex()) {
            throw new Error(`Different script, can not bind contract '${this.constructor.name}' to this UTXO: ${JSON.stringify(utxo)}!`);
        }
        this.utxo = { ...utxo, script: this.lockingScript.toHex() };
        if (this.constructor.isStateful()) {
            this.state = this.constructor.deserializeState(utxo.data);
        }
        return this;
    }
    /**
     * Checks if the contract has state by verifying if the state object exists and is not empty.
     * @returns {boolean} True if the contract has state, false otherwise.
     */
    static isStateful() {
        const selfClazz = this;
        const artifact = selfClazz.artifact;
        if (!artifact) {
            throw new Error(`Artifact is not loaded for the contract: ${this.name}`);
        }
        const stateType = selfClazz.stateType;
        return !!stateType;
    }
}
exports.SmartContract = SmartContract;
SmartContract.newFromCreate = false;
SmartContract.tags = [];
