// @opcat-labs/scrypt-ts-opcat/dist/cjs/psbt/contextProvider.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextProvider = void 0;
const common_js_1 = require('../utils/common.cjs');
const preimage_js_1 = require('../utils/preimage.cjs');
const opcat_1 = require("@opcat-labs/opcat");
const hashes_js_1 = require('../smart-contract/fns/hashes.cjs');
const byteString_js_1 = require('../smart-contract/fns/byteString.cjs');
const index_js_1 = require('../smart-contract/builtin-libs/index.cjs');
const bufferutils_js_1 = require('./bufferutils.cjs');
const tools = __importStar(require("uint8array-tools"));
/** @ignore */
class ContextProvider {
    constructor(psbt) {
        this._curPsbt = psbt;
        this._inputContexts = new Map();
    }
    hasInputCtx(inputIndex) {
        return this._inputContexts.has(inputIndex);
    }
    getInputCtx(inputIndex) {
        if (!this._inputContexts.has(inputIndex)) {
            this.calculateInputCtxs();
        }
        if (!this._inputContexts.has(inputIndex)) {
            throw new Error(`The context of input[${inputIndex}] is not set. Call calculateInputCtxs() first!`);
        }
        return this._inputContexts.get(inputIndex);
    }
    calculateInputCtxs() {
        let spentDataHashes = (0, byteString_js_1.toByteString)('');
        let spentScriptHashes = (0, byteString_js_1.toByteString)('');
        let spentAmounts = (0, byteString_js_1.toByteString)('');
        //let spentDatas: SpentDatas = toByteString('');
        this._curPsbt.data.inputs.forEach((input, inputIndex) => {
            const opcatUtxo = this._curPsbt.getInputOutput(inputIndex);
            const { script, data, value } = opcatUtxo;
            if (script) {
                spentScriptHashes += (0, hashes_js_1.sha256)((0, common_js_1.uint8ArrayToHex)(script));
                spentDataHashes += (0, hashes_js_1.sha256)((0, common_js_1.uint8ArrayToHex)(data));
                spentAmounts += index_js_1.StdUtils.uint64ToByteString(value);
            }
        });
        // // get prevouts for all inputs
        let prevouts = (0, byteString_js_1.toByteString)('');
        this._curPsbt.txInputs.forEach((input) => {
            const hash = typeof input.hash === 'string'
                ? (0, bufferutils_js_1.reverseBuffer)(tools.fromHex(input.hash))
                : input.hash;
            prevouts += `${(0, common_js_1.uint8ArrayToHex)(hash)}${index_js_1.StdUtils.uint32ToByteString(BigInt(input.index))}`;
        });
        const inputs = this._curPsbt.data.inputs
            .map((_, inputIndex) => {
            return {
                inputIndex,
            };
        });
        const preimages = this.calculateInputSHPreimages();
        // calculate input context for input which has tapLeafScript
        // cache input context at this._inputContexts
        inputs.forEach((inputTapLeafHash, index) => {
            const { inputIndex } = inputTapLeafHash;
            const prevout = {
                txHash: (0, common_js_1.uint8ArrayToHex)(this._curPsbt.txInputs[inputIndex].hash),
                outputIndex: BigInt(this._curPsbt.txInputs[inputIndex].index),
            };
            const shPreimage = preimages[index];
            this._inputContexts.set(inputIndex, {
                // ParamCtx
                shPreimage,
                prevouts,
                spentScriptHashes,
                spentAmounts,
                spentDataHashes,
                // DerivedCtx
                prevout,
                inputCount: BigInt(this._curPsbt.txInputs.length),
            });
        });
    }
    calculateInputSHPreimages() {
        return this._curPsbt.unsignedTx.inputs.map((_, inputIndex) => {
            const sigHashType = this._curPsbt.getSigHashType(inputIndex);
            const sighash = sigHashType !== undefined ? sigHashType : opcat_1.crypto.Signature.SIGHASH_ALL;
            const rawSHPreimage = this._curPsbt.unsignedTx.getPreimage(inputIndex, sighash);
            return (0, preimage_js_1.decodeSHPreimage)(new Uint8Array(rawSHPreimage));
        });
    }
}
exports.ContextProvider = ContextProvider;
