// @opcat-labs/scrypt-ts-opcat/dist/cjs/psbt/psbtutils.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.signatureBlocksAction = exports.checkInputForSig = exports.pubkeyInScript = exports.pubkeyPositionInScript = void 0;
const tools = __importStar(require("uint8array-tools"));
const opcat_1 = require("@opcat-labs/opcat");
const signatureutils = __importStar(require('./signatureutils.cjs'));
const hashes_js_1 = require('../smart-contract/fns/hashes.cjs');
/**
 * Finds the position of a public key in a script.
 * @param pubkey The public key to search for.
 * @param script The script to search in.
 * @returns The index of the public key in the script, or -1 if not found.
 * @throws {Error} If there is an unknown script error.
 */
function pubkeyPositionInScript(pubkey, script) {
    const pubkeyHash = tools.fromHex((0, hashes_js_1.hash160)(tools.toHex(pubkey)));
    const decompiled = opcat_1.Script.fromBuffer(Buffer.from(script));
    return decompiled.chunks.findIndex(chunk => {
        return (tools.compare(pubkey, chunk.buf) === 0 ||
            tools.compare(pubkeyHash, chunk.buf) === 0);
    });
}
exports.pubkeyPositionInScript = pubkeyPositionInScript;
/**
 * Checks if a public key is present in a script.
 * @param pubkey The public key to check.
 * @param script The script to search in.
 * @returns A boolean indicating whether the public key is present in the script.
 */
function pubkeyInScript(pubkey, script) {
    return pubkeyPositionInScript(pubkey, script) !== -1;
}
exports.pubkeyInScript = pubkeyInScript;
/**
 * Checks if an input contains a signature for a specific action.
 * @param input - The input to check.
 * @param action - The action to check for.
 * @returns A boolean indicating whether the input contains a signature for the specified action.
 */
function checkInputForSig(input, action) {
    const pSigs = extractPartialSigs(input);
    return pSigs.some(pSig => signatureBlocksAction(pSig, signatureutils.decode, action));
}
exports.checkInputForSig = checkInputForSig;
/**
 * Determines if a given action is allowed for a signature block.
 * @param signature - The signature block.
 * @param signatureDecodeFn - The function used to decode the signature.
 * @param action - The action to be checked.
 * @returns True if the action is allowed, false otherwise.
 */
function signatureBlocksAction(signature, signatureDecodeFn, action) {
    const { hashType } = signatureDecodeFn(signature);
    const whitelist = [];
    const isAnyoneCanPay = hashType & opcat_1.crypto.Signature.SIGHASH_ANYONECANPAY;
    if (isAnyoneCanPay)
        whitelist.push('addInput');
    const hashMod = hashType & 0x1f;
    switch (hashMod) {
        case opcat_1.crypto.Signature.SIGHASH_ALL:
            break;
        case opcat_1.crypto.Signature.SIGHASH_SINGLE:
        case opcat_1.crypto.Signature.SIGHASH_NONE:
            whitelist.push('addOutput');
            whitelist.push('setInputSequence');
            break;
    }
    if (whitelist.indexOf(action) === -1) {
        return true;
    }
    return false;
}
exports.signatureBlocksAction = signatureBlocksAction;
/**
 * Extracts the signatures from a PsbtInput object.
 * If the input has partial signatures, it returns an array of the signatures.
 * If the input does not have partial signatures, it checks if it has a finalScriptSig or finalScriptWitness.
 * If it does, it extracts the signatures from the final scripts and returns them.
 * If none of the above conditions are met, it returns an empty array.
 *
 * @param input - The PsbtInput object from which to extract the signatures.
 * @returns An array of signatures extracted from the PsbtInput object.
 */
function extractPartialSigs(input) {
    let pSigs = [];
    if ((input.partialSig || []).length === 0) {
        if (!input.finalScriptSig && !input.finalScriptWitness)
            return [];
        pSigs = getPsigsFromInputFinalScripts(input);
    }
    else {
        pSigs = input.partialSig;
    }
    return pSigs.map(p => p.signature);
}
/**
 * Retrieves the partial signatures (Psigs) from the input's final scripts.
 * Psigs are extracted from both the final scriptSig and final scriptWitness of the input.
 * Only canonical script signatures are considered.
 *
 * @param input - The PsbtInput object representing the input.
 * @returns An array of PartialSig objects containing the extracted Psigs.
 */
function getPsigsFromInputFinalScripts(input) {
    const scriptItems = !input.finalScriptSig
        ? []
        : opcat_1.Script.fromBuffer(Buffer.from(input.finalScriptSig)).chunks.map(item => item.buf) || [];
    return scriptItems
        .filter(item => {
        return (Buffer.isBuffer(item) && signatureutils.isCanonicalScriptSignature(item));
    })
        .map(sig => ({ signature: sig }));
}
