// @opcat-labs/scrypt-ts-opcat/dist/cjs/providers/dummyProvider.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DummyProvider = void 0;
const utxoProvider_js_1 = require('./utxoProvider.cjs');
const utils = __importStar(require("@noble/hashes/utils"));
const opcat_1 = require("@opcat-labs/opcat");
const extPsbt_js_1 = require('../psbt/extPsbt.cjs');
/**
 * A DummyProvider is build for test purpose only, it always returns a dummy utxo for `getUtxos` request.
 * @category Provider
 */
class DummyProvider {
    constructor(network = 'opcat-mainnet') {
        this.network = network;
        this.broadcastedTxs = new Map();
        this.spentUTXOs = new Set();
        this.newUTXOs = new Map();
    }
    async getNetwork() {
        return this.network;
    }
    async getUtxos(address, _options) {
        const script = opcat_1.Script.fromAddress(address);
        const dummyTx = new opcat_1.Transaction();
        dummyTx.addInput(new opcat_1.Transaction.Input({
            prevTxId: Buffer.from(utils.randomBytes(32)),
            outputIndex: 0,
            script: opcat_1.Script.empty(),
            sequenceNumber: 0xffffffff,
            output: new opcat_1.Transaction.Output({
                script: script,
                satoshis: 2147483647, // 2**31 - 1
                data: '',
            })
        }));
        dummyTx.addOutput(new opcat_1.Transaction.Output({
            script: script,
            satoshis: 210000000,
            data: '',
        }));
        this.broadcastedTxs.set(dummyTx.id, dummyTx.toHex());
        return Promise.resolve([
            {
                txId: dummyTx.id,
                outputIndex: 0,
                script: dummyTx.outputs[0].script.toHex(),
                satoshis: dummyTx.outputs[0].satoshis,
                data: '',
            },
        ]);
    }
    markSpent(txId, vout) {
        const key = `${txId}:${vout}`;
        if (this.newUTXOs.has(key)) {
            this.newUTXOs.delete(key);
        }
        this.spentUTXOs.add(key);
    }
    addNewUTXO(utxo) {
        this.newUTXOs.set((0, utxoProvider_js_1.getUtxoKey)(utxo), utxo);
    }
    getFeeRate() {
        return Promise.resolve(1);
    }
    async getConfirmations(_txId) {
        return Promise.resolve(1);
    }
    async broadcast(txHex) {
        const tx = opcat_1.Transaction.fromString(txHex);
        this.broadcastedTxs.set(tx.id, txHex);
        return tx.id;
    }
    async broadcastPsbt(psbtBase64, metadata) {
        const psbt = extPsbt_js_1.ExtPsbt.fromBase64(psbtBase64);
        const txHex = psbt.extractTransaction().toHex();
        return this.broadcast(txHex);
    }
    async getRawTransaction(txId) {
        const txHex = this.broadcastedTxs.get(txId);
        if (!txHex) {
            throw new Error(`Can not find the tx with id ${txId}, please broadcast it by using the TestProvider first`);
        }
        return txHex;
    }
    async getMedianTime() {
        // Return current timestamp minus 10 minutes (600 seconds)
        return Math.floor(Date.now() / 1000) - 600;
    }
}
exports.DummyProvider = DummyProvider;
