// @opcat-labs/scrypt-ts-opcat/dist/cjs/providers/rpcProvider.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RPCProvider = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
const decimal_js_1 = __importDefault(require("decimal.js"));
const utxoProvider_js_1 = require('./utxoProvider.cjs');
const tools = __importStar(require("uint8array-tools"));
const common_js_1 = require('../utils/common.cjs');
const opcat_1 = require("@opcat-labs/opcat");
const extPsbt_js_1 = require('../psbt/extPsbt.cjs');
/**
 * The RPCProvider is backed by opcat RPC
 * @category Provider
 */
class RPCProvider {
    constructor(network, url, walletName, username, password) {
        this.network = network;
        this.url = url;
        this.walletName = walletName;
        this.username = username;
        this.password = password;
        this.broadcastedTxs = new Map();
        this.spentUTXOs = new Set();
        this.newUTXOs = new Map();
        this.getRpcUser = () => {
            return this.username;
        };
        this.getRpcPassword = () => {
            return this.password;
        };
        this.getRpcUrl = (walletName) => {
            return walletName === null ? this.url : `${this.url}/wallet/${walletName}`;
        };
    }
    async getNetwork() {
        return this.network;
    }
    getFeeRate() {
        const Authorization = `Basic ${tools.toBase64(tools.fromUtf8(`${this.getRpcUser()}:${this.getRpcPassword()}`))}`;
        return fetch(this.getRpcUrl(null), {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization,
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: 'cat-cli',
                method: 'estimatesmartfee',
                params: [1],
            }),
        })
            .then((res) => {
            const contentType = res.headers.get('content-type');
            if (contentType.includes('json')) {
                return res.json();
            }
            else {
                throw new Error(`invalid http content type : ${contentType}, status: ${res.status}`);
            }
        })
            .then((res) => {
            if (res.result === null || (res.result.errors && res.result.errors.length > 0)) {
                throw new Error(JSON.stringify(res));
            }
            const feerate = new decimal_js_1.default(res.result.feerate)
                .mul(new decimal_js_1.default(100000000))
                .div(new decimal_js_1.default(1000))
                .toNumber();
            return Math.ceil(feerate);
        })
            .catch((e) => {
            return 1;
        });
    }
    async getConfirmations(txId) {
        const res = await this._getConfirmations(txId);
        if (res instanceof Error) {
            throw new Error(`getConfirmations failed, ${res.message}`);
        }
        return res.confirmations;
    }
    async _broadcast(txHex) {
        const Authorization = `Basic ${tools.toBase64(tools.fromUtf8(`${this.getRpcUser()}:${this.getRpcPassword()}`))}`;
        return fetch(this.getRpcUrl(null), {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization,
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: 'cat-cli',
                method: 'sendrawtransaction',
                params: [txHex],
            }),
        })
            .then((res) => {
            const contentType = res.headers.get('content-type');
            if (contentType.includes('json')) {
                return res.json();
            }
            else {
                throw new Error(`invalid http content type : ${contentType}, status: ${res.status}`);
            }
        })
            .then((res) => {
            if (res.result === null) {
                throw new Error(JSON.stringify(res));
            }
            return res.result;
        })
            .catch((e) => {
            console.error('sendrawtransaction error:', e);
            return e;
        });
    }
    async _getConfirmations(txid) {
        const Authorization = `Basic ${tools.toBase64(tools.fromUtf8(`${this.getRpcUser()}:${this.getRpcPassword()}`))}`;
        return fetch(this.getRpcUrl(null), {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization,
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: 'cat-cli',
                method: 'getrawtransaction',
                params: [txid, true],
            }),
        })
            .then((res) => {
            const contentType = res.headers.get('content-type');
            if (contentType.includes('json')) {
                return res.json();
            }
            else {
                throw new Error(`invalid http content type : ${contentType}, status: ${res.status}`);
            }
        })
            .then((res) => {
            if (res.result === null) {
                throw new Error(JSON.stringify(res));
            }
            return {
                confirmations: -1,
                blockhash: '',
                ...res.result,
            };
        })
            .catch((e) => {
            console.error('getConfirmations error:', e);
            return e;
        });
    }
    async broadcast(txHex) {
        const res = await this._broadcast(txHex);
        if (res instanceof Error) {
            throw res;
        }
        this.broadcastedTxs.set(res, txHex);
        return res;
    }
    async broadcastPsbt(psbtBase64, metadata) {
        const psbt = extPsbt_js_1.ExtPsbt.fromBase64(psbtBase64);
        const txHex = psbt.extractTransaction().toHex();
        return this.broadcast(txHex);
    }
    async getRawTransaction(txId) {
        let txHex = this.broadcastedTxs.get(txId);
        if (!txHex) {
            const res = await this._getRawTransaction(txId);
            if (res instanceof Error) {
                throw new Error(`Can not find the tx with id ${txId}, please broadcast it first`);
            }
            txHex = res;
        }
        return txHex;
    }
    async _getRawTransaction(txid) {
        const Authorization = `Basic ${tools.toBase64(tools.fromUtf8(`${this.getRpcUser()}:${this.getRpcPassword()}`))}`;
        return fetch(this.getRpcUrl(null), {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization,
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: 'cat-cli',
                method: 'getrawtransaction',
                params: [txid],
            }),
        })
            .then((res) => {
            const contentType = res.headers.get('content-type');
            if (contentType.includes('json')) {
                return res.json();
            }
            else {
                throw new Error(`invalid http content type : ${contentType}, status: ${res.status}`);
            }
        })
            .then((res) => {
            if (res.result === null) {
                throw new Error(JSON.stringify(res));
            }
            return res.result;
        })
            .catch((e) => {
            return e;
        });
    }
    async getUtxos(address, _options) {
        const script = (0, common_js_1.uint8ArrayToHex)(opcat_1.Script.fromAddress(address).toBuffer());
        const Authorization = `Basic ${tools.toBase64(tools.fromUtf8(`${this.getRpcUser()}:${this.getRpcPassword()}`))}`;
        const utxos = await fetch(this.getRpcUrl(this.walletName), {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization,
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: 'cat-cli',
                method: 'listunspent',
                params: [0, 9999999, [address]],
            }),
        })
            .then((res) => {
            if (res.status === 200) {
                return res.json();
            }
            throw new Error(res.statusText);
        })
            .then((res) => {
            if (res.result === null) {
                throw new Error(JSON.stringify(res));
            }
            const utxos = res.result.map((item) => {
                return {
                    txId: item.txid,
                    outputIndex: item.vout,
                    script: item.scriptPubKey,
                    satoshis: new decimal_js_1.default(item.amount).mul(new decimal_js_1.default(100000000)).toNumber(),
                    data: item.data
                };
            });
            return utxos;
        })
            .catch((e) => {
            console.error('listunspent error:', e);
            return [];
        });
        return utxos
            .concat(Array.from(this.newUTXOs.values()))
            .filter((utxo) => this.isUnSpent(utxo.txId, utxo.outputIndex))
            .filter((0, common_js_1.duplicateFilter)((utxo) => `${utxo.txId}:${utxo.outputIndex}`))
            .filter(utxo => utxo.script === script)
            .sort((a, b) => a.satoshis - b.satoshis);
    }
    isUnSpent(txId, vout) {
        const key = `${txId}:${vout}`;
        return !this.spentUTXOs.has(key);
    }
    markSpent(txId, vout) {
        const key = `${txId}:${vout}`;
        if (this.newUTXOs.has(key)) {
            this.newUTXOs.delete(key);
        }
        this.spentUTXOs.add(key);
    }
    addNewUTXO(utxo) {
        this.newUTXOs.set((0, utxoProvider_js_1.getUtxoKey)(utxo), utxo);
    }
    async getMedianTime() {
        const Authorization = `Basic ${tools.toBase64(tools.fromUtf8(`${this.getRpcUser()}:${this.getRpcPassword()}`))}`;
        return fetch(this.getRpcUrl(null), {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Authorization,
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                id: 'cat-cli',
                method: 'getblockchaininfo',
                params: [],
            }),
        })
            .then((res) => {
            const contentType = res.headers.get('content-type');
            if (contentType.includes('json')) {
                return res.json();
            }
            else {
                throw new Error(`invalid http content type : ${contentType}, status: ${res.status}`);
            }
        })
            .then((res) => {
            if (res.result === null) {
                throw new Error(JSON.stringify(res));
            }
            if (typeof res.result.mediantime !== 'number') {
                throw new Error(`Invalid blockchain info: mediantime not found`);
            }
            return res.result.mediantime;
        })
            .catch((e) => {
            throw new Error(`Failed to get median time: ${e.message}`);
        });
    }
}
exports.RPCProvider = RPCProvider;
