// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/builtin-libs/byteStringReader.cjs
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ByteStringReader = void 0;
const decorators_js_1 = require('../decorators.cjs');
const index_js_1 = require('../fns/index.cjs');
const smartContractLib_js_1 = require('../smartContractLib.cjs');
const stdUtils_js_1 = require('./stdUtils.cjs');
/**
 * A reader to parse a ByteString buffer
 * @category Standard Contracts
 */
class ByteStringReader extends smartContractLib_js_1.SmartContractLib {
    constructor(buf) {
        super(buf);
        this.buf = buf;
        this.pos = 0n;
    }
    /**
     * Check if all have been read
     * @returns true if all have been read
     */
    eof() {
        return this.pos >= (0, index_js_1.len)(this.buf);
    }
    /**
     * read bytes which encoded with bitcoin [value-pushing words]{@link https://wiki.bitcoinsv.io/index.php/Opcodes_used_in_Bitcoin_Script}
     * @returns true if all have been read
     */
    readBytes() {
        let l = 0n;
        const buf = this.buf;
        let ret = (0, index_js_1.toByteString)('');
        const header = (0, index_js_1.byteStringToInt)((0, index_js_1.slice)(this.buf, this.pos, this.pos + 1n));
        this.pos++;
        if (header < stdUtils_js_1.StdUtils.OP_PUSHDATA1_VAL) {
            l = header;
            ret = (0, index_js_1.slice)(buf, this.pos, this.pos + l);
        }
        else if (header == stdUtils_js_1.StdUtils.OP_PUSHDATA1_VAL) {
            l = stdUtils_js_1.StdUtils.fromLEUnsigned((0, index_js_1.slice)(buf, this.pos, this.pos + 1n));
            this.pos += 1n;
            ret = (0, index_js_1.slice)(this.buf, this.pos, this.pos + l);
        }
        else if (header == stdUtils_js_1.StdUtils.OP_PUSHDATA2_VAL) {
            l = stdUtils_js_1.StdUtils.fromLEUnsigned((0, index_js_1.slice)(this.buf, this.pos, this.pos + 2n));
            this.pos += 2n;
            ret = (0, index_js_1.slice)(this.buf, this.pos, this.pos + l);
        }
        else if (header == stdUtils_js_1.StdUtils.OP_PUSHDATA4_VAL) {
            l = stdUtils_js_1.StdUtils.fromLEUnsigned((0, index_js_1.slice)(this.buf, this.pos, this.pos + 4n));
            this.pos += 4n;
            ret = (0, index_js_1.slice)(this.buf, this.pos, this.pos + l);
        }
        else {
            // shall not reach here
            (0, index_js_1.assert)(false);
        }
        this.pos += l;
        return ret;
    }
    /**
     * read a byte as boolean
     * @returns true if the read byte not equal to '00'
     */
    readBool() {
        const buf = (0, index_js_1.slice)(this.buf, this.pos, this.pos + 1n);
        this.pos++;
        return (0, index_js_1.toByteString)('00') != buf;
    }
    /**
     * read bytes as `readBytes` and convert it to a number with `byteString2Int`
     * @returns a number
     */
    readVarint() {
        let ret = (0, index_js_1.toByteString)('');
        const header = (0, index_js_1.slice)(this.buf, this.pos, this.pos + 1n);
        this.pos++;
        if (header == stdUtils_js_1.StdUtils.VARINT_2BYTE) {
            ret = (0, index_js_1.slice)(this.buf, this.pos, this.pos + 2n);
            this.pos += 2n;
        }
        else if (header == stdUtils_js_1.StdUtils.VARINT_4BYTE) {
            ret = (0, index_js_1.slice)(this.buf, this.pos, this.pos + 4n);
            this.pos += 4n;
        }
        else if (header == stdUtils_js_1.StdUtils.VARINT_8BYTE) {
            ret = (0, index_js_1.slice)(this.buf, this.pos, this.pos + 8n);
            this.pos += 8n;
        }
        else {
            ret = header;
        }
        return stdUtils_js_1.StdUtils.fromLEUnsigned(ret);
    }
}
exports.ByteStringReader = ByteStringReader;
__decorate([
    (0, decorators_js_1.prop)(),
    __metadata("design:type", String)
], ByteStringReader.prototype, "buf", void 0);
__decorate([
    (0, decorators_js_1.prop)(),
    __metadata("design:type", BigInt)
], ByteStringReader.prototype, "pos", void 0);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Boolean)
], ByteStringReader.prototype, "eof", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", String)
], ByteStringReader.prototype, "readBytes", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Boolean)
], ByteStringReader.prototype, "readBool", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", BigInt)
], ByteStringReader.prototype, "readVarint", null);
