// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/builtin-libs/p2pkh.cjs
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.P2PKH = void 0;
const decorators_js_1 = require('../decorators.cjs');
const assert_js_1 = require('../fns/assert.cjs');
const primitives_js_1 = require('../types/primitives.cjs');
const smartContract_js_1 = require('../smartContract.cjs');
const pubKey2Addr_js_1 = require('../fns/pubKey2Addr.cjs');
/**
 * A Pay-to-Public-Key-Hash (P2PKH) smart contract that allows spending
 * by providing a signature matching the specified public key hash.
 *
 * @property addr - The address derived from the public key hash that can unlock this contract
 * @method unlock - Verifies the provided signature matches the contract's public key hash
 * @param sig - The signature to verify
 * @param pubkey - The public key corresponding to the address
 * @throws If public key hash doesn't match or signature verification fails
 */
class P2PKH extends smartContract_js_1.SmartContract {
    constructor(addr) {
        super(...arguments);
        this.addr = addr;
    }
    unlock(sig, pubkey) {
        (0, assert_js_1.assert)((0, pubKey2Addr_js_1.pubKey2Addr)(pubkey) == this.addr, 'public key hashes are not equal');
        (0, assert_js_1.assert)(this.checkSig(sig, pubkey), 'signature check failed');
    }
}
exports.P2PKH = P2PKH;
__decorate([
    (0, decorators_js_1.prop)(),
    __metadata("design:type", String)
], P2PKH.prototype, "addr", void 0);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, String]),
    __metadata("design:returntype", void 0)
], P2PKH.prototype, "unlock", null);
const desc = {
    version: 9,
    compilerVersion: '1.19.0+commit.72eaeba',
    contract: 'P2PKH',
    md5: '0c046dfb1f1a91cf72b9a852537bdfe1',
    structs: [],
    library: [],
    alias: [],
    abi: [
        {
            type: 'function',
            name: 'unlock',
            index: 0,
            params: [
                {
                    name: 'sig',
                    type: 'Sig',
                },
                {
                    name: 'pubkey',
                    type: 'PubKey',
                },
            ],
        },
        {
            type: 'constructor',
            params: [
                {
                    name: 'addr',
                    type: 'Ripemd160',
                },
            ],
        },
    ],
    stateProps: [],
    buildType: 'release',
    file: '',
    hex: '76a9<addr>88ac',
    sourceMapFile: '',
};
P2PKH.loadArtifact(desc);
