// @opcat-labs/scrypt-ts-opcat/dist/esm/smart-contract/types/primitives.js
import { getValidatedHexString } from './utils.js';
/**
 * Converts a ByteString to a validated SigHashPreimage.
 * @param b - The ByteString to convert
 * @returns The validated SigHashPreimage
 */
export function SigHashPreimage(b) {
    return getValidatedHexString(b, false);
}
/**
 * Creates a `PrivKey` instance from a `Int32`.
 * @category Global Functions
 * @onchain
 * @param key - Input Int32.
 * @returns - A domain specific private key representation.
 */
export function PrivKey(n) {
    return n;
}
/**
 * Creates a `PubKey` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific public key representation.
 */
export function PubKey(b) {
    return getValidatedHexString(b, false);
}
/**
 * Creates a `Sig` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific digital signature representation.
 */
export function Sig(b) {
    return getValidatedHexString(b, false);
}
/**
 * Creates a `Ripemd160` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific RIPEMD-160 hash representation.
 */
export function Ripemd160(b) {
    return getValidatedHexString(b, false);
}
/**
 * Creates a `PubKeyHash` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific address representation.
 */
export function PubKeyHash(b) {
    return Ripemd160(b);
}
/**
 * Creates an `Addr` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific address representation.
 */
export function Addr(b) {
    return PubKeyHash(b);
}
/**
 * Creates a `Sha1` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific SHA-1 hash representation.
 */
export function Sha1(b) {
    return getValidatedHexString(b, false);
}
/**
 * Creates a `Sha256` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific SHA-256 hash representation.
 */
export function Sha256(b) {
    return getValidatedHexString(b, false);
}
/**
 * Creates a `OpCodeType` from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific OpCodeType representation.
 */
export function OpCodeType(b) {
    return getValidatedHexString(b, false);
}
OpCodeType.prototype.toNumber = function () {
    return parseInt(this, 16);
};
