// @opcat-labs/scrypt-ts-opcat/dist/esm/smart-contract/types/abi.js
/**
 * @ignore
 */
export var ScryptType;
(function (ScryptType) {
    ScryptType["BOOL"] = "bool";
    ScryptType["INT"] = "int";
    ScryptType["BYTES"] = "bytes";
    ScryptType["PUBKEY"] = "PubKey";
    ScryptType["PRIVKEY"] = "PrivKey";
    ScryptType["SIG"] = "Sig";
    ScryptType["RIPEMD160"] = "Ripemd160";
    ScryptType["PubKeyHash"] = "PubKeyHash";
    ScryptType["SHA1"] = "Sha1";
    ScryptType["SHA256"] = "Sha256";
    ScryptType["SIGHASHTYPE"] = "SigHashType";
    ScryptType["SIGHASHPREIMAGE"] = "SigHashPreimage";
    ScryptType["OPCODETYPE"] = "OpCodeType";
})(ScryptType || (ScryptType = {}));
/**
 * @ignore
 */
export var SymbolType;
(function (SymbolType) {
    SymbolType["ScryptType"] = "ScryptType";
    SymbolType["Contract"] = "Contract";
    SymbolType["Library"] = "Library";
    SymbolType["Struct"] = "Struct";
    SymbolType["Unknown"] = "Unknown";
})(SymbolType || (SymbolType = {}));
/**
 * @ignore
 */
export function isScryptType(type) {
    return Object.keys(ScryptType)
        .map((key) => ScryptType[key])
        .includes(type);
}
/**
 * @ignore
 */
export function isSubBytes(type) {
    return [
        ScryptType.OPCODETYPE,
        ScryptType.PUBKEY,
        ScryptType.RIPEMD160,
        ScryptType.SHA1,
        ScryptType.SHA256,
        ScryptType.SIG,
        ScryptType.SIGHASHPREIMAGE,
        ScryptType.SIGHASHTYPE,
        ScryptType.PubKeyHash,
    ]
        .map((t) => t.toString())
        .includes(type);
}
/**
 * @ignore
 */
export function isBytes(type) {
    return type === ScryptType.BYTES || isSubBytes(type);
}
