// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/fns/bitwise.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.xor = exports.or = exports.and = exports.invert = exports.rshift = exports.lshift = void 0;
const assert_js_1 = require('./assert.cjs');
const byteString_js_1 = require('./byteString.cjs');
const tools = __importStar(require("uint8array-tools"));
/**
 * Logical left shift matching on-chain OP_LSHIFT semantics.
 *
 * On-chain, OP_LSHIFT treats the stack item as raw bytes (sign-magnitude
 * encoding), interprets those bytes as an unsigned big-endian integer,
 * performs an unsigned logical left shift, and truncates the result back to
 * the original byte length.  This function replicates that behaviour so that
 * off-chain tests produce the same result as the deployed contract.
 *
 * Decision: the previous arithmetic-multiplication implementation diverged
 * from on-chain semantics for any negative or large value; replaced with a
 * byte-level unsigned shift + truncation to match OP_LSHIFT exactly.
 *
 * More detail about [Bitwise Operations]{@link https://docs.opcatlabs.io/how-to-write-a-contract/built-ins#bitwise-operator}
 * @category Bitwise Operations
 */
function lshift(x, n) {
    (0, assert_js_1.assert)(n >= 0n, 'n < 0');
    const hex = (0, byteString_js_1.intToByteString)(x);
    if (hex.length === 0 || n === 0n)
        return x;
    const buf = tools.fromHex(hex);
    const origLen = buf.length;
    // Treat raw bytes as unsigned big-endian integer and shift left
    const unsigned = BigInt('0x' + tools.toHex(buf));
    const shifted = unsigned << n;
    // Truncate to original byte length: take last origLen bytes
    const shiftedHex = shifted.toString(16);
    const paddedHex = shiftedHex.length < origLen * 2
        ? shiftedHex.padStart(origLen * 2, '0')
        : shiftedHex;
    const truncatedHex = paddedHex.slice(-origLen * 2);
    return (0, byteString_js_1.byteStringToInt)(truncatedHex);
}
exports.lshift = lshift;
/**
 * Logical right shift matching on-chain OP_RSHIFT semantics.
 *
 * On-chain, OP_RSHIFT treats the stack item as raw bytes (sign-magnitude
 * encoding), interprets those bytes as an unsigned big-endian integer,
 * performs an unsigned logical right shift, and truncates the result back to
 * the original byte length.  This function replicates that behaviour so that
 * off-chain tests produce the same result as the deployed contract.
 *
 * Decision: the previous arithmetic-division implementation diverged from
 * on-chain semantics for any negative value; replaced with a byte-level
 * unsigned shift + truncation to match OP_RSHIFT exactly.
 *
 * Example: rshift(-6n, 2n)
 *   -6 in sign-magnitude = 0x86; unsigned = 134; 134 >>> 2 = 33 = 0x21
 *   result = byteStringToInt('21') = 33n  (not -1n as the old code returned)
 *
 * More detail about [Bitwise Operations]{@link https://docs.opcatlabs.io/how-to-write-a-contract/built-ins#bitwise-operator}
 * @category Bitwise Operations
 */
function rshift(x, n) {
    (0, assert_js_1.assert)(n >= 0n, 'n < 0');
    const hex = (0, byteString_js_1.intToByteString)(x);
    if (hex.length === 0 || n === 0n)
        return x;
    const buf = tools.fromHex(hex);
    const origLen = buf.length;
    // Treat raw bytes as unsigned big-endian integer and shift right
    const unsigned = BigInt('0x' + tools.toHex(buf));
    const shifted = unsigned >> n;
    // Truncate to original byte length: take last origLen bytes
    const shiftedHex = shifted.toString(16).padStart(origLen * 2, '0');
    const truncatedHex = shiftedHex.slice(-origLen * 2);
    return (0, byteString_js_1.byteStringToInt)(truncatedHex);
}
exports.rshift = rshift;
/**
 * Inverts the bits of an Int value.
 * If the input value is 0n, it returns 0n directly.
 * @category Bitwise Operations
 * @onchain
 * @param a - The Int value to be inverted.
 * @returns {Int32} The inverted Int value.
 */
function invert(a) {
    if (a === 0n) {
        return a;
    }
    const buffer = tools.fromHex((0, byteString_js_1.intToByteString)(a));
    for (let i = 0; i < buffer.length; i++) {
        buffer[i] = ~buffer[i];
    }
    return (0, byteString_js_1.byteStringToInt)(tools.toHex(buffer));
}
exports.invert = invert;
/**
 * Performs a bitwise AND operation between two Int values.
 * @category Bitwise Operations
 * @onchain
 * @param a - First integer value
 * @param b - Second integer value
 * @returns Result of bitwise AND operation as Int
 */
function and(a, b) {
    const size1 = (0, byteString_js_1.intToByteString)(a).length / 2;
    const size2 = (0, byteString_js_1.intToByteString)(b).length / 2;
    const maxSize = BigInt(Math.max(size1, size2));
    const ba = tools.fromHex((0, byteString_js_1.intToByteString)(a, maxSize));
    const bb = tools.fromHex((0, byteString_js_1.intToByteString)(b, maxSize));
    for (let i = 0; i < ba.length; i++) {
        ba[i] &= bb[i];
    }
    return (0, byteString_js_1.byteStringToInt)(tools.toHex(ba));
}
exports.and = and;
/**
 * Performs a bitwise OR operation on two Int values.
 * @category Bitwise Operations
 * @onchain
 * @param a - First integer value
 * @param b - Second integer value
 * @returns Result of bitwise OR operation as Int
 */
function or(a, b) {
    const size1 = (0, byteString_js_1.intToByteString)(a).length / 2;
    const size2 = (0, byteString_js_1.intToByteString)(b).length / 2;
    const maxSize = BigInt(Math.max(size1, size2));
    const ba = tools.fromHex((0, byteString_js_1.intToByteString)(a, maxSize));
    const bb = tools.fromHex((0, byteString_js_1.intToByteString)(b, maxSize));
    for (let i = 0; i < ba.length; i++) {
        ba[i] |= bb[i];
    }
    return (0, byteString_js_1.byteStringToInt)(tools.toHex(ba));
}
exports.or = or;
/**
 * Performs a bitwise XOR operation on two Int values.
 * @category Bitwise Operations
 * @onchain
 * @param a First integer value
 * @param b Second integer value
 * @returns Result of the XOR operation as an Int
 */
function xor(a, b) {
    const size1 = (0, byteString_js_1.intToByteString)(a).length / 2;
    const size2 = (0, byteString_js_1.intToByteString)(b).length / 2;
    const maxSize = BigInt(Math.max(size1, size2));
    const ba = tools.fromHex((0, byteString_js_1.intToByteString)(a, maxSize));
    const bb = tools.fromHex((0, byteString_js_1.intToByteString)(b, maxSize));
    for (let i = 0; i < ba.length; i++) {
        ba[i] ^= bb[i];
    }
    return (0, byteString_js_1.byteStringToInt)(tools.toHex(ba));
}
exports.xor = xor;
