// @opcat-labs/scrypt-ts-opcat/dist/esm/utils/common.js
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Address } from '@opcat-labs/opcat';
import * as tools from 'uint8array-tools';
/**
 * Converts a Uint8Array to a hexadecimal string.
 * @param {Uint8Array} bytes - The input array of bytes.
 * @returns {string} The hexadecimal representation of the input bytes.
 */
export function uint8ArrayToHex(bytes) {
    return tools.toHex(bytes);
}
/**
 * Converts a hexadecimal string to a Uint8Array.
 * @param {string} hex - The hexadecimal string.
 * @returns {Uint8Array} The resulting Uint8Array.
 */
export function hexToUint8Array(hex) {
    if (hex.length % 2 !== 0) {
        throw new Error('Hex string must have an even length');
    }
    return tools.fromHex(hex);
}
/**
 * create a FixedArray and fill it with initial value
 * @param value
 * @param len
 * @returns
 */
export function fillFixedArray(value, len) {
    const array = new Array(len);
    array.fill(value);
    return array;
}
/** @ignore */
export function requireTrue(res, message) {
    if (!res) {
        throw new Error(message);
    }
}
/** @ignore */
export function getTxId(input) {
    return uint8ArrayToHex(input.prevTxId);
}
/** @ignore */
export function cloneDeep(obj, hash = new WeakMap()) {
    if (obj === null || typeof obj !== 'object') {
        return obj;
    }
    if (hash.has(obj)) {
        return hash.get(obj);
    }
    if (obj instanceof Date) {
        return new Date(obj.getTime());
    }
    if (obj instanceof RegExp) {
        return new RegExp(obj);
    }
    if (Buffer.isBuffer(obj)) {
        return Buffer.from(Uint8Array.prototype.slice.call(obj));
    }
    if (ArrayBuffer.isView(obj) && !(obj instanceof DataView)) {
        const constructor = obj.constructor;
        return new constructor(obj.buffer.slice(0));
    }
    if (obj instanceof ArrayBuffer) {
        return obj.slice(0);
    }
    if (obj instanceof DataView) {
        const buffer = obj.buffer.slice(0);
        return new DataView(buffer, obj.byteOffset, obj.byteLength);
    }
    if (obj instanceof Map) {
        const copy = new Map();
        hash.set(obj, copy);
        obj.forEach((value, key) => {
            copy.set(cloneDeep(key, hash), cloneDeep(value, hash));
        });
        return copy;
    }
    if (obj instanceof Set) {
        const copy = new Set();
        hash.set(obj, copy);
        obj.forEach((value) => {
            copy.add(cloneDeep(value, hash));
        });
        return copy;
    }
    if (Array.isArray(obj)) {
        const copy = [];
        hash.set(obj, copy);
        obj.forEach((item, index) => {
            copy[index] = cloneDeep(item, hash);
        });
        return copy;
    }
    const copy = {};
    hash.set(obj, copy);
    Object.setPrototypeOf(copy, Object.getPrototypeOf(obj));
    const allProps = [...Object.getOwnPropertyNames(obj), ...Object.getOwnPropertySymbols(obj)];
    allProps.forEach((key) => {
        const descriptor = Object.getOwnPropertyDescriptor(obj, key);
        if (descriptor) {
            if (descriptor.get || descriptor.set) {
                Object.defineProperty(copy, key, descriptor);
            }
            else {
                Object.defineProperty(copy, key, {
                    ...descriptor,
                    value: cloneDeep(descriptor.value, hash),
                });
            }
        }
    });
    return copy;
}
/**
 * convert satoshi to hex in little-endian order
 * @param value
 * @returns
 */
export function satoshiToHex(value) {
    const bf = new Uint8Array(8);
    tools.writeUInt64(bf, 0, value, 'LE');
    return uint8ArrayToHex(bf);
}
/**
 * convert utf8 text to hex string
 * @param text
 * @returns
 */
export function textToHex(text) {
    return uint8ArrayToHex(tools.fromUtf8(text));
}
export function duplicateFilter(uniqueFn) {
    return function (value, index, arr) {
        const uniqueArr = arr.map(uniqueFn);
        const uniqueIndex = uniqueArr.findIndex((t) => t === uniqueFn(value));
        return uniqueIndex === index;
    };
}
const MAXINT = 0xffffffff; // Math.pow(2, 32) - 1;
export function isFinal(sequenceNumber) {
    return sequenceNumber !== MAXINT;
}
export function toHex(val) {
    if (Buffer.isBuffer(val)) {
        return val.toString('hex');
    }
    if (typeof val === 'string') {
        return val;
    }
    return tools.toHex(val);
}
/**
 * Converts a Opcat address to its public key hash (PKH) in hex format.
 * @param address - The Opcat address to convert.
 * @returns The public key hash as a hex string.
 */
export function addrToPkh(address) {
    return tools.toHex(Address.fromString(address).hashBuffer);
}
export function pushData(data) {
    const res = [];
    const dLen = data.length;
    if (dLen < 0x4c) {
        const dLenBuff = Buffer.alloc(1);
        dLenBuff.writeUInt8(dLen);
        res.push(dLenBuff);
    }
    else if (dLen <= 0xff) {
        // OP_PUSHDATA1
        res.push(Buffer.from('4c', 'hex'));
        const dLenBuff = Buffer.alloc(1);
        dLenBuff.writeUInt8(dLen);
        res.push(dLenBuff);
    }
    else if (dLen <= 0xffff) {
        // OP_PUSHDATA2
        res.push(Buffer.from('4d', 'hex'));
        const dLenBuff = Buffer.alloc(2);
        dLenBuff.writeUint16LE(dLen);
        res.push(dLenBuff);
    }
    else {
        // OP_PUSHDATA4
        res.push(Buffer.from('4e', 'hex'));
        const dLenBuff = Buffer.alloc(4);
        dLenBuff.writeUint32LE(dLen);
        res.push(dLenBuff);
    }
    res.push(data);
    return Buffer.concat(res);
}
export function splitChunks(buffer, chunkSize) {
    const chunks = [];
    for (let i = 0; i < buffer.length; i += chunkSize) {
        chunks.push(buffer.slice(i, i + chunkSize));
    }
    return chunks;
}
