// @opcat-labs/scrypt-ts-opcat/dist/esm/utils/preimage.js
import * as tools from 'uint8array-tools';
import { byteStringToInt, intToByteString, toByteString } from '../smart-contract/fns/byteString.js';
/**
 * Decodes a serialized SHPreimage from a Uint8Array.
 *
 * @param preimage - The serialized preimage data as a Uint8Array
 * @returns The decoded SHPreimage object containing:
 *          - Version number
 *          - Hash of previous outputs
 *          - Spent script hash
 *          - Spent data hash
 *          - Spent amount
 *          - Sequence number
 *          - Hash of spent amounts
 *          - Hash of spent scripts
 *          - Hash of spent datas
 *          - Hash of sequences
 *          - Hash of outputs
 *          - Input index
 *          - Lock time
 *          - Sighash type
 */
export function decodeSHPreimage(preimage) {
    return {
        nVersion: toByteString(tools.toHex(preimage.subarray(0, 4))),
        hashPrevouts: tools.toHex(preimage.subarray(4, 36)),
        inputIndex: byteStringToInt(tools.toHex(preimage.subarray(36, 40))),
        outpoint: tools.toHex(preimage.subarray(40, 76)),
        spentScriptHash: tools.toHex(preimage.subarray(76, 108)),
        spentDataHash: tools.toHex(preimage.subarray(108, 140)),
        value: byteStringToInt(tools.toHex(preimage.subarray(140, 148))),
        nSequence: toByteString(tools.toHex(preimage.subarray(148, 152))),
        hashSpentAmounts: tools.toHex(preimage.subarray(152, 184)),
        hashSpentScriptHashes: tools.toHex(preimage.subarray(184, 216)),
        hashSpentDataHashes: tools.toHex(preimage.subarray(216, 248)),
        hashSequences: tools.toHex(preimage.subarray(248, 280)),
        hashOutputs: tools.toHex(preimage.subarray(280, 312)),
        nLockTime: byteStringToInt(tools.toHex(preimage.subarray(312, 316))),
        sigHashType: byteStringToInt(tools.toHex(preimage.subarray(316, 320))),
    };
}
/**
 * Encodes a SHPreimage object into a ByteString by concatenating its fields.
 * Each field is converted to a ByteString with appropriate formatting.
 *
 * @param shPreimage - The SHPreimage object containing all required fields
 * @returns The concatenated ByteString representation of the SHPreimage
 */
export function encodeSHPreimage(shPreimage) {
    const rawSHPreimage = shPreimage.nVersion
        + shPreimage.hashPrevouts
        + intToByteString(shPreimage.inputIndex, 4n)
        + shPreimage.outpoint
        + shPreimage.spentScriptHash
        + shPreimage.spentDataHash
        + intToByteString(shPreimage.value, 8n)
        + shPreimage.nSequence
        + shPreimage.hashSpentAmounts
        + shPreimage.hashSpentScriptHashes
        + shPreimage.hashSpentDataHashes
        + shPreimage.hashSequences
        + shPreimage.hashOutputs
        + intToByteString(shPreimage.nLockTime, 4n)
        + intToByteString(shPreimage.sigHashType, 4n);
    return rawSHPreimage;
}
