// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/builtin-libs/txUtils.cjs
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxUtils = void 0;
const assert_js_1 = require('../fns/assert.cjs');
const consts_js_1 = require('../consts.cjs');
const decorators_js_1 = require('../decorators.cjs');
const index_js_1 = require('../fns/index.cjs');
const opCode_js_1 = require('../types/opCode.cjs');
const smartContractLib_js_1 = require('../smartContractLib.cjs');
const index_js_2 = require('../types/index.cjs');
const stdUtils_js_1 = require('./stdUtils.cjs');
/**
 * Library for parsing and constructing transactions
 * @category Library
 * @onchain
 */
class TxUtils extends smartContractLib_js_1.SmartContractLib {
    /**
     * Build serialized tx output
     * @param script serialized locking script of the output
     * @param satoshis serialized satoshis of the output
     * @returns serialized tx output in format ByteString
     */
    static buildOutput(scriptHash, satoshis) {
        return TxUtils.buildDataOutput(scriptHash, satoshis, (0, index_js_1.sha256)((0, index_js_1.toByteString)('')));
    }
    /**
   * Build serialized tx output
   * @param script serialized locking script of the output
   * @param satoshis serialized satoshis of the output
   * @returns serialized tx output in format ByteString
   */
    static buildDataOutput(scriptHash, satoshis, dataHash) {
        const scriptHashLen = (0, index_js_1.len)(scriptHash);
        const dataHashLen = (0, index_js_1.len)(dataHash);
        //const dataLen = len(data);
        (0, assert_js_1.assert)(scriptHashLen == 32n, "script hash length must be equal to 32");
        (0, assert_js_1.assert)(dataHashLen == 32n, "data hash length must be equal to 32");
        (0, assert_js_1.assert)(satoshis > 0n, "satoshis must be greater than 0");
        return TxUtils.satoshisToByteString(satoshis) + scriptHash + dataHash;
    }
    /**
     * Build serialized change output
     * @param change change output to build
     * @returns serialized change output in format ByteString
     */
    static buildChangeOutput(change) {
        return change.satoshis > 0n
            ? TxUtils.buildDataOutput(change.scriptHash, change.satoshis, change.dataHash)
            : (0, index_js_1.toByteString)('');
    }
    /**
     * Merge tx input into a ByteString
     * @param txInput tx input, must be a segwit input
     * @returns serialized tx input
     */
    static mergeInput(txInput) {
        (0, assert_js_1.assert)((0, index_js_1.len)(txInput.prevTxHash) == consts_js_1.TX_INPUT_PREV_TX_HASH_BYTE_LEN);
        (0, assert_js_1.assert)((0, index_js_1.len)(txInput.scriptHash) == consts_js_1.TX_INPUT_SCRIPT_HASH_BYTE_LEN);
        return (txInput.prevTxHash +
            stdUtils_js_1.StdUtils.uint32ToByteString(txInput.prevOutputIndex) +
            txInput.scriptHash +
            stdUtils_js_1.StdUtils.uint32ToByteString(txInput.sequence));
    }
    /**
     * build `OP_RETURN` script from data payload
     * @param {ByteString} data the data payload
     * @returns {ByteString} a ByteString contains the data payload
     */
    static buildOpReturnOutput(data) {
        const script = (0, index_js_1.toByteString)('6a') + (0, index_js_1.intToByteString)((0, index_js_1.len)(data)) + data;
        return TxUtils.satoshisToByteString(TxUtils.ZERO_SATS) + (0, index_js_1.sha256)(script) + (0, index_js_1.sha256)((0, index_js_1.toByteString)(''));
    }
    /**
     * constructs a P2PKH script from a given PubKeyHash
     * @param {PubKeyHash} pubKeyHash - the recipient's public key hash
     * @returns {ByteString} a `ByteString` representing the P2PKH script
     */
    static buildP2PKHScript(addr) {
        return ((0, index_js_1.toByteString)(opCode_js_1.OpCode.OP_DUP) +
            (0, index_js_1.toByteString)(opCode_js_1.OpCode.OP_HASH160) +
            (0, index_js_1.intToByteString)(20n) +
            addr +
            (0, index_js_1.toByteString)(opCode_js_1.OpCode.OP_EQUALVERIFY) +
            (0, index_js_1.toByteString)(opCode_js_1.OpCode.OP_CHECKSIG));
    }
    /**
     * constructs a P2PKH output from a given PubKeyHash and satoshi amount
     * @param {Addr} addr - the recipient's public key hash
     * @param {Int32} amount - the satoshi amount
     * @returns {ByteString} a `ByteString` representing the P2PKH output
     */
    static buildP2PKHOutput(amount, addr) {
        return TxUtils.buildDataOutput((0, index_js_1.sha256)(TxUtils.buildP2PKHScript(addr)), amount, (0, index_js_1.sha256)((0, index_js_1.toByteString)('')));
    }
    /**
     * build `OP_FALSE OP_RETURN` script from data payload
     * @param {ByteString} data the data payload
     * @returns {ByteString} a ByteString contains the data payload
     */
    static buildOpreturnScript(data) {
        return (0, index_js_1.toByteString)(opCode_js_1.OpCode.OP_FALSE) + (0, index_js_1.toByteString)(opCode_js_1.OpCode.OP_RETURN) + stdUtils_js_1.StdUtils.pushData(data);
    }
    /**
     * convert a `UInt64` number to 8 bytes in little-end order.
     * @param {UInt64} n - the satoshi amount
     * @returns {ByteString} a `ByteString`
     */
    static satoshisToByteString(n) {
        return stdUtils_js_1.StdUtils.uint64ToByteString(n);
    }
    /**
     * convert a `ByteString` to a `UInt64` number
     * @param {ByteString} bs - the satoshi amount
     * @returns {UInt64} a `UInt64`
     */
    static byteStringToSatoshis(bs) {
        (0, assert_js_1.assert)((0, index_js_1.len)(bs) == 8n, "satoshis must be 8 bytes");
        return stdUtils_js_1.StdUtils.fromLEUnsigned(bs);
    }
}
exports.TxUtils = TxUtils;
/** if a output satoshi value is zero */
TxUtils.ZERO_SATS = 0n;
__decorate([
    (0, decorators_js_1.prop)(),
    __metadata("design:type", BigInt)
], TxUtils, "ZERO_SATS", void 0);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, BigInt]),
    __metadata("design:returntype", String)
], TxUtils, "buildOutput", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, BigInt, String]),
    __metadata("design:returntype", String)
], TxUtils, "buildDataOutput", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", String)
], TxUtils, "buildChangeOutput", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", String)
], TxUtils, "mergeInput", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", String)
], TxUtils, "buildOpReturnOutput", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", String)
], TxUtils, "buildP2PKHScript", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [BigInt, String]),
    __metadata("design:returntype", String)
], TxUtils, "buildP2PKHOutput", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", String)
], TxUtils, "buildOpreturnScript", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [BigInt]),
    __metadata("design:returntype", String)
], TxUtils, "satoshisToByteString", null);
__decorate([
    (0, decorators_js_1.method)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", BigInt)
], TxUtils, "byteStringToSatoshis", null);
