// @opcat-labs/scrypt-ts-opcat/dist/esm/smart-contract/builtin-libs/byteStringReader.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { method, prop } from '../decorators.js';
import { assert, byteStringToInt, len, slice, toByteString } from '../fns/index.js';
import { SmartContractLib } from '../smartContractLib.js';
import { StdUtils } from './stdUtils.js';
/**
 * A reader to parse a ByteString buffer
 * @category Standard Contracts
 */
export class ByteStringReader extends SmartContractLib {
    constructor(buf) {
        super(buf);
        this.buf = buf;
        this.pos = 0n;
    }
    /**
     * Check if all have been read
     * @returns true if all have been read
     */
    eof() {
        return this.pos >= len(this.buf);
    }
    /**
     * read bytes which encoded with bitcoin [value-pushing words]{@link https://wiki.bitcoinsv.io/index.php/Opcodes_used_in_Bitcoin_Script}
     * @returns true if all have been read
     */
    readBytes() {
        let l = 0n;
        const buf = this.buf;
        let ret = toByteString('');
        const header = byteStringToInt(slice(this.buf, this.pos, this.pos + 1n));
        this.pos++;
        if (header < StdUtils.OP_PUSHDATA1_VAL) {
            l = header;
            ret = slice(buf, this.pos, this.pos + l);
        }
        else if (header == StdUtils.OP_PUSHDATA1_VAL) {
            l = StdUtils.fromLEUnsigned(slice(buf, this.pos, this.pos + 1n));
            this.pos += 1n;
            ret = slice(this.buf, this.pos, this.pos + l);
        }
        else if (header == StdUtils.OP_PUSHDATA2_VAL) {
            l = StdUtils.fromLEUnsigned(slice(this.buf, this.pos, this.pos + 2n));
            this.pos += 2n;
            ret = slice(this.buf, this.pos, this.pos + l);
        }
        else if (header == StdUtils.OP_PUSHDATA4_VAL) {
            l = StdUtils.fromLEUnsigned(slice(this.buf, this.pos, this.pos + 4n));
            this.pos += 4n;
            ret = slice(this.buf, this.pos, this.pos + l);
        }
        else {
            // shall not reach here
            assert(false);
        }
        this.pos += l;
        return ret;
    }
    /**
     * read a byte as boolean
     * @returns true if the read byte not equal to '00'
     */
    readBool() {
        const buf = slice(this.buf, this.pos, this.pos + 1n);
        this.pos++;
        return toByteString('00') != buf;
    }
    /**
     * read bytes as `readBytes` and convert it to a number with `byteString2Int`
     * @returns a number
     */
    readVarint() {
        let ret = toByteString('');
        const header = slice(this.buf, this.pos, this.pos + 1n);
        this.pos++;
        if (header == StdUtils.VARINT_2BYTE) {
            ret = slice(this.buf, this.pos, this.pos + 2n);
            this.pos += 2n;
        }
        else if (header == StdUtils.VARINT_4BYTE) {
            ret = slice(this.buf, this.pos, this.pos + 4n);
            this.pos += 4n;
        }
        else if (header == StdUtils.VARINT_8BYTE) {
            ret = slice(this.buf, this.pos, this.pos + 8n);
            this.pos += 8n;
        }
        else {
            ret = header;
        }
        return StdUtils.fromLEUnsigned(ret);
    }
}
__decorate([
    prop(),
    __metadata("design:type", String)
], ByteStringReader.prototype, "buf", void 0);
__decorate([
    prop(),
    __metadata("design:type", BigInt)
], ByteStringReader.prototype, "pos", void 0);
__decorate([
    method(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Boolean)
], ByteStringReader.prototype, "eof", null);
__decorate([
    method(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", String)
], ByteStringReader.prototype, "readBytes", null);
__decorate([
    method(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Boolean)
], ByteStringReader.prototype, "readBool", null);
__decorate([
    method(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", BigInt)
], ByteStringReader.prototype, "readVarint", null);
