// @opcat-labs/scrypt-ts-opcat/dist/cjs/utils/contractHeader.cjs
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractHeaderSerializer = exports.CONTRACT_HEADER_FIELD_FLAGS = void 0;
const opcat_1 = require("@opcat-labs/opcat");
const index_js_1 = require('../smart-contract/types/index.cjs');
const index_js_2 = require('../smart-contract/fns/index.cjs');
const cbor2_1 = require("cbor2");
const common_js_1 = require('./common.cjs');
const constants_js_1 = require('./constants.cjs');
/**
 * Fields allowed in header, like ordinals
 */
exports.CONTRACT_HEADER_FIELD_FLAGS = {
    // 20 bytes md5 hash
    // serialize:
    //   serialized = CONTRACT_HEADER_FIELDS.DM5 <hash>
    MD5: index_js_1.OpCode.OP_1,
    // arbitrary user defined tag
    // serialize:
    //    tagParts = cbor.encode(tag)
    //    serialized = CONTRACT_HEADER_FIELDS.TAG + tagParts[0] + CONTRACT_HEADER_FIELDS.TAG + tagParts[1] + ....
    TAGS: index_js_1.OpCode.OP_2,
};
class ContractHeaderSerializer {
    static get ENVELOPE_HEAD_HEX() {
        return this.ENVELOPE.head.join('');
    }
    static get ENVELOPE_TAIL_HEX() {
        return this.ENVELOPE.tail.join('');
    }
    /**
     * serialize the header, and concat the serialized header with the lockingScript without header
     * @param {ContractHeader} header the contract header
     * @param {string} contractLockingScript the lockingScript without header
     * @returns the script that used by transaction output
     */
    static serialize(header, contractLockingScript) {
        const headerHex = this.sealHeader(header);
        return headerHex + contractLockingScript;
    }
    /**
     * deserialize the header and lockingScript from the transaction output script
     * @param txOutScript the transaction output script
     * @returns the header and the contractLockingScript
     */
    static deserialize(txOutScript) {
        return this.unsealHeader(txOutScript);
    }
    static pushField(bufs, field) {
        bufs.push(Buffer.from(exports.CONTRACT_HEADER_FIELD_FLAGS[field], 'hex'));
    }
    static pushCbor(bufs, field, value) {
        if (value === undefined || value === null) {
            return bufs;
        }
        const data = Buffer.from((0, cbor2_1.encode)(value));
        const dataChunks = (0, common_js_1.splitChunks)(Array.from(data), this.LIMIT);
        for (const chunk of dataChunks) {
            this.pushField(bufs, field);
            bufs.push((0, common_js_1.pushData)(Buffer.from(chunk)));
        }
        return bufs;
    }
    static pushShortField(bufs, field, value) {
        if (typeof value !== 'string') {
            throw new Error(`Invalid value type for field ${field}: ${typeof value}, expected string`);
        }
        this.pushField(bufs, field);
        bufs.push((0, common_js_1.pushData)(Buffer.from(value, 'hex')));
    }
    static sealHeader(header) {
        let bufs = [];
        bufs.push(Buffer.from(this.ENVELOPE_HEAD_HEX, 'hex'));
        bufs.push(Buffer.from((0, index_js_2.intToByteString)(header.version, 1n), 'hex'));
        this.pushShortField(bufs, 'MD5', header.md5);
        this.pushCbor(bufs, 'TAGS', header.tags);
        bufs.push(Buffer.from(this.ENVELOPE_TAIL_HEX, 'hex'));
        return Buffer.concat(bufs).toString('hex');
    }
    static unsealHeader(txOutScript) {
        if (!txOutScript.startsWith(this.ENVELOPE_HEAD_HEX)) {
            return {
                header: null,
                lockingScript: txOutScript,
            };
        }
        let version;
        let md5 = '';
        let tags = '';
        txOutScript = txOutScript.slice(this.ENVELOPE_HEAD_HEX.length);
        version = (0, index_js_2.byteStringToInt)(txOutScript.slice(0, 2));
        txOutScript = txOutScript.slice(2);
        const bodyAsm = opcat_1.Script.fromHex(txOutScript).toASM();
        const bodyAsmItems = bodyAsm.split(' ');
        let readIndex = 0;
        while (readIndex < bodyAsmItems.length) {
            const tagOP = opcat_1.Script.fromASM(bodyAsmItems[readIndex]).toHex().toLowerCase();
            if (tagOP == exports.CONTRACT_HEADER_FIELD_FLAGS.MD5.toLowerCase()) {
                md5 = bodyAsmItems[readIndex + 1];
                readIndex += 2;
            }
            else if (tagOP == exports.CONTRACT_HEADER_FIELD_FLAGS.TAGS.toLowerCase()) {
                tags += bodyAsmItems[readIndex + 1];
                readIndex += 2;
            }
            else if (tagOP == index_js_1.OpCode.OP_ENDIF.toLowerCase()) {
                readIndex += 1;
                break;
            }
            else {
                throw new Error(`Invalid contract header OP: ${tagOP}`);
            }
        }
        const lockingScript = opcat_1.Script.fromASM(bodyAsmItems.slice(readIndex).join(' ')).toHex();
        let header = {
            version,
            md5,
            tags: tags.length > 0 ? (0, cbor2_1.decode)(Buffer.from(tags, 'hex')) : [],
        };
        return { header, lockingScript };
    }
}
exports.ContractHeaderSerializer = ContractHeaderSerializer;
_a = ContractHeaderSerializer;
// current version of contract header
ContractHeaderSerializer.VERSION = 1n;
ContractHeaderSerializer.SCRYPT_SYMBOL = '736372';
ContractHeaderSerializer.LIMIT = constants_js_1.MAX_OP_PUSH_DATA_SIZE;
ContractHeaderSerializer.ENVELOPE = {
    head: [
        index_js_1.OpCode.OP_FALSE,
        index_js_1.OpCode.OP_IF,
        '04', // op_push_4_bytes: symbol('scr' 3 bytes) + version(1 byte)
        (0, index_js_2.toByteString)(_a.SCRYPT_SYMBOL),
    ],
    tail: [
        index_js_1.OpCode.OP_ENDIF,
    ]
};
