// @opcat-labs/scrypt-ts-opcat/dist/cjs/utils/sigUtils.cjs
"use strict";
/**
 * Off-chain signing utilities for preimage verification.
 *
 * This module provides functions to generate signatures off-chain using a
 * hardcoded private key. The signatures can then be verified on-chain using
 * checkDataSig, avoiding the need for on-chain signature generation which
 * bloats script size.
 *
 * @module sigUtils
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyPreimageSig = exports.signSHPreimageForCheckDataSig = exports.signData = exports.signDataWithInternalKey = exports.signSHPreimage = exports.signPreimage = void 0;
const opcat_1 = require("@opcat-labs/opcat");
const hashes_js_1 = require('../smart-contract/fns/hashes.cjs');
const primitives_js_1 = require('../smart-contract/types/primitives.cjs');
const contextUtils_js_1 = require('../smart-contract/builtin-libs/contextUtils.cjs');
const preimage_js_1 = require('./preimage.cjs');
const { ECDSA, Signature } = opcat_1.crypto;
/**
 * The internal private key derived from ContextUtils.privKey.
 * This ensures consistency between off-chain signing and on-chain verification.
 */
const INTERNAL_KEY = opcat_1.PrivateKey.fromHex(contextUtils_js_1.ContextUtils.privKey.toString(16).padStart(64, '0'), opcat_1.Networks.defaultNetwork);
/**
 * Signs a serialized preimage off-chain using the internal private key (from ContextUtils).
 *
 * This function generates a DER-encoded ECDSA signature that can be verified
 * on-chain using checkSig. The message is hashed with hash256 (double SHA256)
 * to match the transaction interpreter's sighash calculation.
 *
 * Note: For checkDataSig (which uses single SHA256), a different signature
 * would be needed. The JavaScript runtime uses this hash256 version for
 * compatibility with checkSig verification.
 *
 * @param preimage - The serialized preimage bytes to sign
 * @param sigHashType - The signature hash type (default: 0x01 for SIGHASH_ALL)
 * @returns A DER-encoded signature with sigHashType appended
 */
function signPreimage(preimage, sigHashType = 0x01) {
    // Compute hash256 of the preimage (double SHA256 for checkSig compatibility)
    // Reverse the hash to match checkSigImpl's verification format
    const hash = Buffer.from((0, hashes_js_1.hash256)(preimage), 'hex').reverse();
    // Sign the hash using ECDSA with internal key
    const signature = ECDSA.sign(hash, INTERNAL_KEY, 'little');
    // Get DER encoded signature
    const derSig = signature.toDER();
    // Append sigHashType byte
    const sigHashTypeByte = Buffer.from([sigHashType]);
    const fullSig = Buffer.concat([derSig, sigHashTypeByte]);
    return (0, primitives_js_1.Sig)(fullSig.toString('hex'));
}
exports.signPreimage = signPreimage;
/**
 * Signs a SHPreimage off-chain by first serializing it.
 *
 * This is a convenience wrapper around signPreimage that handles
 * the serialization of the SHPreimage struct.
 *
 * Note: Uses encodeSHPreimage (same as checkSigImpl) to ensure
 * the signature matches what checkSig expects.
 *
 * @param shPreimage - The SHPreimage struct to sign
 * @param sigHashType - The signature hash type (default: 0x01 for SIGHASH_ALL).
 *                      Should match shPreimage.sigHashType for verification to succeed.
 * @returns A DER-encoded signature with sigHashType appended
 */
function signSHPreimage(shPreimage, sigHashType = 0x01) {
    // Serialize the SHPreimage using encodeSHPreimage (same as checkSigImpl uses)
    const preimage = (0, preimage_js_1.encodeSHPreimage)(shPreimage);
    return signPreimage(preimage, sigHashType);
}
exports.signSHPreimage = signSHPreimage;
/**
 * Signs arbitrary data off-chain using the internal key (ContextUtils.privKey).
 *
 * This function generates a pure DER-encoded ECDSA signature (NO sighash type)
 * that can be verified on-chain using checkDataSig. The message is hashed with
 * single SHA256 to match OP_CHECKSIGFROMSTACK behavior.
 *
 * Note: This uses the internal key from ContextUtils. For signing with
 * a custom private key (e.g., Oracle scenarios), use signData() instead.
 *
 * @param message - The message bytes to sign
 * @returns A pure DER-encoded signature (no sighash type appended)
 */
function signDataWithInternalKey(message) {
    return signData(INTERNAL_KEY, message);
}
exports.signDataWithInternalKey = signDataWithInternalKey;
/**
 * Signs arbitrary data with a custom private key for use with checkDataSig (OP_CHECKSIGFROMSTACK).
 *
 * This function generates a pure DER-encoded ECDSA signature (NO sighash type)
 * that can be verified on-chain using checkDataSig. The message is hashed with
 * single SHA256 to match OP_CHECKSIGFROMSTACK behavior.
 *
 * Use this method for Oracle scenarios where you need to sign data with a specific
 * private key that the contract will verify against.
 *
 * @param privateKey - The private key to sign with
 * @param message - The message bytes to sign
 * @returns A pure DER-encoded signature (no sighash type appended)
 *
 * @example
 * ```typescript
 * // Oracle signs price data
 * const oraclePrivKey = PrivateKey.fromWIF('...');
 * const priceData = toByteString('BTC/USD:50000', true);
 * const sig = signData(oraclePrivKey, priceData);
 *
 * // Contract verifies using checkDataSig
 * // assert(this.checkDataSig(sig, priceData, oraclePubKey));
 * ```
 */
function signData(privateKey, message) {
    // Compute single SHA256 of the message for checkDataSig compatibility
    // Reverse the hash for little-endian format (same as OP_CHECKSIGFROMSTACK)
    const hash = Buffer.from((0, hashes_js_1.sha256)(message), 'hex').reverse();
    // Sign the hash using ECDSA
    const signature = ECDSA.sign(hash, privateKey, 'little');
    // Get pure DER encoded signature (NO sighash type for OP_CHECKSIGFROMSTACK)
    const derSig = signature.toDER();
    return (0, primitives_js_1.Sig)(derSig.toString('hex'));
}
exports.signData = signData;
/**
 * Signs a SHPreimage off-chain for use with checkDataSig (OP_CHECKSIGFROMSTACK).
 *
 * This is a convenience wrapper that serializes the SHPreimage and signs it
 * using single SHA256 (matching checkDataSig behavior).
 *
 * @param shPreimage - The SHPreimage struct to sign
 * @returns A pure DER-encoded signature (no sighash type appended)
 */
function signSHPreimageForCheckDataSig(shPreimage) {
    // Serialize the SHPreimage using encodeSHPreimage
    const preimage = (0, preimage_js_1.encodeSHPreimage)(shPreimage);
    return signDataWithInternalKey(preimage);
}
exports.signSHPreimageForCheckDataSig = signSHPreimageForCheckDataSig;
/**
 * Verifies a preimage signature off-chain.
 *
 * This function can be used to verify signatures before submitting
 * transactions, ensuring the signature is valid.
 *
 * @param sig - The signature to verify
 * @param preimage - The preimage that was signed
 * @returns true if the signature is valid, false otherwise
 */
function verifyPreimageSig(sig, preimage) {
    try {
        // Compute SHA256 hash of the preimage
        const hash = Buffer.from((0, hashes_js_1.sha256)(preimage), 'hex');
        // Get the signature bytes (without sigHashType)
        const sigBytes = Buffer.from(sig, 'hex');
        const derSig = sigBytes.slice(0, sigBytes.length - 1);
        // Parse the DER signature
        const signature = Signature.fromDER(derSig);
        // Get public key from ContextUtils
        const publicKey = opcat_1.PublicKey.fromHex(contextUtils_js_1.ContextUtils.pubKey);
        // Verify the signature
        return ECDSA.verify(hash, signature, publicKey, 'little');
    }
    catch (e) {
        return false;
    }
}
exports.verifyPreimageSig = verifyPreimageSig;
