// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/types/abi.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBytes = exports.isSubBytes = exports.isScryptType = exports.SymbolType = exports.ScryptType = void 0;
/**
 * @ignore
 */
var ScryptType;
(function (ScryptType) {
    ScryptType["BOOL"] = "bool";
    ScryptType["INT"] = "int";
    ScryptType["BYTES"] = "bytes";
    ScryptType["PUBKEY"] = "PubKey";
    ScryptType["PRIVKEY"] = "PrivKey";
    ScryptType["SIG"] = "Sig";
    ScryptType["RIPEMD160"] = "Ripemd160";
    ScryptType["PubKeyHash"] = "PubKeyHash";
    ScryptType["SHA1"] = "Sha1";
    ScryptType["SHA256"] = "Sha256";
    ScryptType["SIGHASHTYPE"] = "SigHashType";
    ScryptType["SIGHASHPREIMAGE"] = "SigHashPreimage";
    ScryptType["OPCODETYPE"] = "OpCodeType";
})(ScryptType || (exports.ScryptType = ScryptType = {}));
/**
 * @ignore
 */
var SymbolType;
(function (SymbolType) {
    SymbolType["ScryptType"] = "ScryptType";
    SymbolType["Contract"] = "Contract";
    SymbolType["Library"] = "Library";
    SymbolType["Struct"] = "Struct";
    SymbolType["Unknown"] = "Unknown";
})(SymbolType || (exports.SymbolType = SymbolType = {}));
/**
 * @ignore
 */
function isScryptType(type) {
    return Object.keys(ScryptType)
        .map((key) => ScryptType[key])
        .includes(type);
}
exports.isScryptType = isScryptType;
/**
 * @ignore
 */
function isSubBytes(type) {
    return [
        ScryptType.OPCODETYPE,
        ScryptType.PUBKEY,
        ScryptType.RIPEMD160,
        ScryptType.SHA1,
        ScryptType.SHA256,
        ScryptType.SIG,
        ScryptType.SIGHASHPREIMAGE,
        ScryptType.SIGHASHTYPE,
        ScryptType.PubKeyHash,
    ]
        .map((t) => t.toString())
        .includes(type);
}
exports.isSubBytes = isSubBytes;
/**
 * @ignore
 */
function isBytes(type) {
    return type === ScryptType.BYTES || isSubBytes(type);
}
exports.isBytes = isBytes;
