// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/types/primitives.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpCodeType = exports.Sha256 = exports.Sha1 = exports.Addr = exports.PubKeyHash = exports.Ripemd160 = exports.Sig = exports.PubKey = exports.PrivKey = exports.SigHashPreimage = void 0;
const utils_js_1 = require('./utils.cjs');
/**
 * Converts a ByteString to a validated SigHashPreimage.
 * @param b - The ByteString to convert
 * @returns The validated SigHashPreimage
 */
function SigHashPreimage(b) {
    return (0, utils_js_1.getValidatedHexString)(b, false);
}
exports.SigHashPreimage = SigHashPreimage;
/**
 * Creates a `PrivKey` instance from a `Int32`.
 * @category Global Functions
 * @onchain
 * @param key - Input Int32.
 * @returns - A domain specific private key representation.
 */
function PrivKey(n) {
    return n;
}
exports.PrivKey = PrivKey;
/**
 * Creates a `PubKey` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific public key representation.
 */
function PubKey(b) {
    return (0, utils_js_1.getValidatedHexString)(b, false);
}
exports.PubKey = PubKey;
/**
 * Creates a `Sig` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific digital signature representation.
 */
function Sig(b) {
    return (0, utils_js_1.getValidatedHexString)(b, false);
}
exports.Sig = Sig;
/**
 * Creates a `Ripemd160` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific RIPEMD-160 hash representation.
 */
function Ripemd160(b) {
    return (0, utils_js_1.getValidatedHexString)(b, false);
}
exports.Ripemd160 = Ripemd160;
/**
 * Creates a `PubKeyHash` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific address representation.
 */
function PubKeyHash(b) {
    return Ripemd160(b);
}
exports.PubKeyHash = PubKeyHash;
/**
 * Creates an `Addr` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific address representation.
 */
function Addr(b) {
    return PubKeyHash(b);
}
exports.Addr = Addr;
/**
 * Creates a `Sha1` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific SHA-1 hash representation.
 */
function Sha1(b) {
    return (0, utils_js_1.getValidatedHexString)(b, false);
}
exports.Sha1 = Sha1;
/**
 * Creates a `Sha256` instance from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific SHA-256 hash representation.
 */
function Sha256(b) {
    return (0, utils_js_1.getValidatedHexString)(b, false);
}
exports.Sha256 = Sha256;
/**
 * Creates a `OpCodeType` from a `ByteString`.
 * @category Global Functions
 * @onchain
 * @param b - Input ByteString.
 * @returns - A domain specific OpCodeType representation.
 */
function OpCodeType(b) {
    return (0, utils_js_1.getValidatedHexString)(b, false);
}
exports.OpCodeType = OpCodeType;
OpCodeType.prototype.toNumber = function () {
    return parseInt(this, 16);
};
