// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/deserializer.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeArgfromHex = exports.hex2bytes = exports.hex2bool = exports.hex2int = void 0;
const abiutils_js_1 = require('./abiutils.cjs');
const byteString_js_1 = require('./fns/byteString.cjs');
const abi_js_1 = require('./types/abi.cjs');
const script_js_1 = require('./types/script.cjs');
function hex2int(hex) {
    if (hex === '00') {
        return BigInt(0);
    }
    else if (hex === '4f') {
        return BigInt(-1);
    }
    else {
        const b = script_js_1.Script.fromHex(hex);
        const chuck = b.chunks[0];
        if (chuck.opcodenum >= 81 && chuck.opcodenum <= 96) {
            return BigInt(chuck.opcodenum - 80);
        }
        return (0, byteString_js_1.byteStringToInt)(chuck.buf.toString('hex'));
    }
}
exports.hex2int = hex2int;
function hex2bool(hex) {
    if (hex === '51') {
        return true;
    }
    else if (hex === '00') {
        return false;
    }
    throw new Error(`invalid hex ${hex}`);
}
exports.hex2bool = hex2bool;
function hex2bytes(hex) {
    if (hex === '00') {
        return '';
    }
    const s = script_js_1.Script.fromHex(hex);
    const chuck = s.chunks[0];
    if (chuck.opcodenum >= 81 && chuck.opcodenum <= 96) {
        return Buffer.from([chuck.opcodenum - 80]).toString('hex');
    }
    return chuck.buf.toString('hex');
}
exports.hex2bytes = hex2bytes;
function deserializer(type, hex) {
    switch (type) {
        case abi_js_1.ScryptType.BOOL:
            return hex2bool(hex);
        case abi_js_1.ScryptType.INT:
        case abi_js_1.ScryptType.PRIVKEY:
            return BigInt(hex2int(hex));
        // case ScryptType.SIGHASHTYPE:
        //   return Number(hex2int(hex)) as SigHashType;
        default:
            if ((0, abi_js_1.isBytes)(type)) {
                return hex2bytes(hex);
            }
            throw new Error(`<${type}> cannot be cast to ScryptType, only sCrypt native types supported`);
    }
}
function createStruct(resolver, param, opcodesMap) {
    const structTypeInfo = resolver(param.type);
    const entity = structTypeInfo.info;
    const obj = Object.create({});
    entity.params.forEach(p => {
        const typeInfo = resolver(p.type);
        if ((0, abiutils_js_1.isArrayType)(typeInfo.finalType)) {
            Object.assign(obj, {
                [p.name]: createArray(resolver, typeInfo.finalType, `${param.name}.${p.name}`, opcodesMap)
            });
        }
        else if (typeInfo.symbolType === abi_js_1.SymbolType.Struct) {
            Object.assign(obj, {
                [p.name]: createStruct(resolver, { name: `${param.name}.${p.name}`, type: p.type }, opcodesMap)
            });
        }
        else if (typeInfo.symbolType === abi_js_1.SymbolType.Library) {
            Object.assign(obj, {
                [p.name]: createLibrary(resolver, { name: `${param.name}.${p.name}`, type: p.type }, opcodesMap)
            });
        }
        else {
            Object.assign(obj, {
                [p.name]: deserializer(typeInfo.finalType, opcodesMap.get(`<${param.name}.${p.name}>`))
            });
        }
    });
    return obj;
}
function createLibrary(resolver, param, opcodesMap) {
    const libraryTypeInfo = resolver(param.type);
    const entity = libraryTypeInfo.info;
    return entity.params.map(p => {
        const typeInfo = resolver(p.type);
        if ((0, abiutils_js_1.isArrayType)(typeInfo.finalType)) {
            return createArray(resolver, typeInfo.finalType, `${param.name}.${p.name}`, opcodesMap);
        }
        else if (typeInfo.symbolType === abi_js_1.SymbolType.Struct) {
            return createStruct(resolver, { name: `${param.name}.${p.name}`, type: p.type }, opcodesMap);
        }
        else if (typeInfo.symbolType === abi_js_1.SymbolType.Library) {
            return createLibrary(resolver, { name: `${param.name}.${p.name}`, type: p.type }, opcodesMap);
        }
        else {
            return deserializer(typeInfo.finalType, opcodesMap.get(`<${param.name}.${p.name}>`));
        }
    });
}
function createArray(resolver, type, name, opcodesMap) {
    const arrays = [];
    const [elemTypeName, sizes] = (0, abiutils_js_1.arrayTypeAndSize)(type);
    const arraylen = sizes[0];
    if (sizes.length === 1) {
        for (let index = 0; index < arraylen; index++) {
            const typeInfo = resolver(elemTypeName);
            if (typeInfo.symbolType === abi_js_1.SymbolType.Struct) {
                arrays.push(createStruct(resolver, {
                    name: `${name}[${index}]`,
                    type: typeInfo.finalType
                }, opcodesMap));
            }
            else if (typeInfo.symbolType === abi_js_1.SymbolType.Library) {
                arrays.push(createLibrary(resolver, {
                    name: `${name}[${index}]`,
                    type: typeInfo.finalType
                }, opcodesMap));
            }
            else {
                arrays.push(deserializer(typeInfo.finalType, opcodesMap.get(`<${name}[${index}]>`)));
            }
        }
    }
    else {
        for (let index = 0; index < arraylen; index++) {
            const finalType = resolver(elemTypeName).finalType;
            const subArrayType = [finalType, sizes.slice(1).map(size => `[${size}]`).join('')].join('');
            arrays.push(createArray(resolver, subArrayType, `${name}[${index}]`, opcodesMap));
        }
    }
    return arrays;
}
function deserializeArgfromHex(resolver, arg, opcodesMap) {
    let value;
    const typeInfo = resolver(arg.type);
    if ((0, abiutils_js_1.isArrayType)(typeInfo.finalType)) {
        value = createArray(resolver, arg.type, arg.name, opcodesMap);
    }
    else if (typeInfo.symbolType === abi_js_1.SymbolType.Struct) {
        value = createStruct(resolver, arg, opcodesMap);
    }
    else if (typeInfo.symbolType === abi_js_1.SymbolType.Library) {
        value = createLibrary(resolver, arg, opcodesMap);
    }
    else {
        value = deserializer(arg.type, opcodesMap.get(`<${arg.name}>`));
    }
    arg.value = value;
    return arg;
}
exports.deserializeArgfromHex = deserializeArgfromHex;
