// @opcat-labs/scrypt-ts-opcat/dist/esm/smart-contract/builtin-libs/p2pkh.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { method, prop } from "../decorators.js";
import { assert } from "../fns/assert.js";
import { Addr, PubKey, Sig } from "../types/primitives.js";
import { SmartContract } from "../smartContract.js";
import { pubKey2Addr } from "../fns/pubKey2Addr.js";
/**
 * A Pay-to-Public-Key-Hash (P2PKH) smart contract that allows spending
 * by providing a signature matching the specified public key hash.
 *
 * @property addr - The address derived from the public key hash that can unlock this contract
 * @method unlock - Verifies the provided signature matches the contract's public key hash
 * @param sig - The signature to verify
 * @param pubkey - The public key corresponding to the address
 * @throws If public key hash doesn't match or signature verification fails
 */
export class P2PKH extends SmartContract {
    constructor(addr) {
        super(...arguments);
        this.addr = addr;
    }
    unlock(sig, pubkey) {
        assert(pubKey2Addr(pubkey) == this.addr, 'public key hashes are not equal');
        assert(this.checkSig(sig, pubkey), 'signature check failed');
    }
}
__decorate([
    prop(),
    __metadata("design:type", String)
], P2PKH.prototype, "addr", void 0);
__decorate([
    method(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, String]),
    __metadata("design:returntype", void 0)
], P2PKH.prototype, "unlock", null);
const desc = {
    version: 9,
    compilerVersion: '1.19.0+commit.72eaeba',
    contract: 'P2PKH',
    md5: '0c046dfb1f1a91cf72b9a852537bdfe1',
    structs: [],
    library: [],
    alias: [],
    abi: [
        {
            type: 'function',
            name: 'unlock',
            index: 0,
            params: [
                {
                    name: 'sig',
                    type: 'Sig',
                },
                {
                    name: 'pubkey',
                    type: 'PubKey',
                },
            ],
        },
        {
            type: 'constructor',
            params: [
                {
                    name: 'addr',
                    type: 'Ripemd160',
                },
            ],
        },
    ],
    stateProps: [],
    buildType: 'release',
    file: '',
    hex: '76a9<addr>88ac',
    sourceMapFile: '',
};
P2PKH.loadArtifact(desc);
