// @opcat-labs/scrypt-ts-opcat/dist/cjs/smart-contract/fns/hashes.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ripemd160 = exports.sha1 = exports.hash256 = exports.sha256 = exports.hash160 = void 0;
const sha2_1 = require("@noble/hashes/sha2");
const legacy_1 = require("@noble/hashes/legacy");
const index_js_1 = require('../types/index.cjs');
const common_js_1 = require('../../utils/common.cjs');
/**
 * A RIPEMD160 hash of a SHA256 hash, which is always 160 bits or 20 bytes long.
 * This value is commonly used inside Bitcoin, particularly for Bitcoin
 * addresses.
 *
 * See:
 * https://en.wikipedia.org/wiki/RIPEMD
 * @category Hashing
 * @onchain
 * @param {ByteString} a ByteString Data, a.k.a. pre-image, which can be any size.
 * @returns {Ripemd160} The hash in the form of a string.
 */
function hash160(a) {
    return (0, common_js_1.uint8ArrayToHex)((0, legacy_1.ripemd160)((0, sha2_1.sha256)((0, common_js_1.hexToUint8Array)(a))));
}
exports.hash160 = hash160;
/**
 * A SHA256 hash, which is always 256 bits or 32 bytes long.
 *
 * See:
 * https://www.movable-type.co.uk/scripts/sha256.html
 * @category Hashing
 * @onchain
 * @param {ByteString} a ByteString Data, a.k.a. pre-image, which can be any size.
 * @returns {Sha256} The hash in the form of a string.
 */
function sha256(a) {
    return (0, common_js_1.uint8ArrayToHex)((0, sha2_1.sha256)((0, common_js_1.hexToUint8Array)(a)));
}
exports.sha256 = sha256;
/**
 * A double SHA256 hash, which is always 256 bits or 32 bytes bytes long. This
 * hash function is commonly used inside Bitcoin, particularly for the hash of a
 * block and the hash of a transaction.
 *
 * See:
 * https://www.movable-type.co.uk/scripts/sha256.html
 * @category Hashing
 * @onchain
 * @param {ByteString} a ByteString data, a.k.a. pre-image, which can be any size.
 * @returns {Sha256} The hash in the form of a string.
 */
function hash256(a) {
    return sha256(sha256(a));
}
exports.hash256 = hash256;
/**
 * A SHA or SHA1 hash, which is always 160 bits or 20 bytes long.
 *
 * See:
 * https://en.wikipedia.org/wiki/SHA-1
 * @category Hashing
 * @onchain
 * @param {ByteString} a ByteString Data, a.k.a. pre-image, which can be any size.
 * @returns {Sha1} The hash in the form of a string.
 */
function sha1(a) {
    return (0, index_js_1.Sha1)((0, common_js_1.uint8ArrayToHex)((0, legacy_1.sha1)((0, common_js_1.hexToUint8Array)(a))));
}
exports.sha1 = sha1;
/**
 * A RIPEMD160 hash, which is always 160 bits or 20 bytes long.
 * See:
 * https://en.wikipedia.org/wiki/RIPEMD
 * @category Hashing
 * @onchain
 * @param {ByteString} a ByteString Data, a.k.a. pre-image, which can be any size.
 * @returns {Ripemd160} The hash in the form of a ByteString.
 */
function ripemd160(a) {
    return (0, index_js_1.Ripemd160)((0, common_js_1.uint8ArrayToHex)((0, legacy_1.ripemd160)((0, common_js_1.hexToUint8Array)(a))));
}
exports.ripemd160 = ripemd160;
