// @opcat-labs/scrypt-ts-opcat/dist/cjs/utils/common.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.splitChunks = exports.pushData = exports.addrToPkh = exports.toHex = exports.isFinal = exports.duplicateFilter = exports.textToHex = exports.satoshiToHex = exports.cloneDeep = exports.getTxId = exports.requireTrue = exports.fillFixedArray = exports.hexToUint8Array = exports.uint8ArrayToHex = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */
const opcat_1 = require("@opcat-labs/opcat");
const tools = __importStar(require("uint8array-tools"));
/**
 * Converts a Uint8Array to a hexadecimal string.
 * @param {Uint8Array} bytes - The input array of bytes.
 * @returns {string} The hexadecimal representation of the input bytes.
 */
function uint8ArrayToHex(bytes) {
    return tools.toHex(bytes);
}
exports.uint8ArrayToHex = uint8ArrayToHex;
/**
 * Converts a hexadecimal string to a Uint8Array.
 * @param {string} hex - The hexadecimal string.
 * @returns {Uint8Array} The resulting Uint8Array.
 */
function hexToUint8Array(hex) {
    if (hex.length % 2 !== 0) {
        throw new Error('Hex string must have an even length');
    }
    return tools.fromHex(hex);
}
exports.hexToUint8Array = hexToUint8Array;
/**
 * create a FixedArray and fill it with initial value
 * @param value
 * @param len
 * @returns
 */
function fillFixedArray(value, len) {
    const array = new Array(len);
    array.fill(value);
    return array;
}
exports.fillFixedArray = fillFixedArray;
/** @ignore */
function requireTrue(res, message) {
    if (!res) {
        throw new Error(message);
    }
}
exports.requireTrue = requireTrue;
/** @ignore */
function getTxId(input) {
    return uint8ArrayToHex(input.prevTxId);
}
exports.getTxId = getTxId;
/** @ignore */
function cloneDeep(obj, hash = new WeakMap()) {
    if (obj === null || typeof obj !== 'object') {
        return obj;
    }
    if (hash.has(obj)) {
        return hash.get(obj);
    }
    if (obj instanceof Date) {
        return new Date(obj.getTime());
    }
    if (obj instanceof RegExp) {
        return new RegExp(obj);
    }
    if (Buffer.isBuffer(obj)) {
        return Buffer.from(Uint8Array.prototype.slice.call(obj));
    }
    if (ArrayBuffer.isView(obj) && !(obj instanceof DataView)) {
        const constructor = obj.constructor;
        return new constructor(obj.buffer.slice(0));
    }
    if (obj instanceof ArrayBuffer) {
        return obj.slice(0);
    }
    if (obj instanceof DataView) {
        const buffer = obj.buffer.slice(0);
        return new DataView(buffer, obj.byteOffset, obj.byteLength);
    }
    if (obj instanceof Map) {
        const copy = new Map();
        hash.set(obj, copy);
        obj.forEach((value, key) => {
            copy.set(cloneDeep(key, hash), cloneDeep(value, hash));
        });
        return copy;
    }
    if (obj instanceof Set) {
        const copy = new Set();
        hash.set(obj, copy);
        obj.forEach((value) => {
            copy.add(cloneDeep(value, hash));
        });
        return copy;
    }
    if (Array.isArray(obj)) {
        const copy = [];
        hash.set(obj, copy);
        obj.forEach((item, index) => {
            copy[index] = cloneDeep(item, hash);
        });
        return copy;
    }
    const copy = {};
    hash.set(obj, copy);
    Object.setPrototypeOf(copy, Object.getPrototypeOf(obj));
    const allProps = [...Object.getOwnPropertyNames(obj), ...Object.getOwnPropertySymbols(obj)];
    allProps.forEach((key) => {
        const descriptor = Object.getOwnPropertyDescriptor(obj, key);
        if (descriptor) {
            if (descriptor.get || descriptor.set) {
                Object.defineProperty(copy, key, descriptor);
            }
            else {
                Object.defineProperty(copy, key, {
                    ...descriptor,
                    value: cloneDeep(descriptor.value, hash),
                });
            }
        }
    });
    return copy;
}
exports.cloneDeep = cloneDeep;
/**
 * convert satoshi to hex in little-endian order
 * @param value
 * @returns
 */
function satoshiToHex(value) {
    const bf = new Uint8Array(8);
    tools.writeUInt64(bf, 0, value, 'LE');
    return uint8ArrayToHex(bf);
}
exports.satoshiToHex = satoshiToHex;
/**
 * convert utf8 text to hex string
 * @param text
 * @returns
 */
function textToHex(text) {
    return uint8ArrayToHex(tools.fromUtf8(text));
}
exports.textToHex = textToHex;
function duplicateFilter(uniqueFn) {
    return function (value, index, arr) {
        const uniqueArr = arr.map(uniqueFn);
        const uniqueIndex = uniqueArr.findIndex((t) => t === uniqueFn(value));
        return uniqueIndex === index;
    };
}
exports.duplicateFilter = duplicateFilter;
const MAXINT = 0xffffffff; // Math.pow(2, 32) - 1;
function isFinal(sequenceNumber) {
    return sequenceNumber !== MAXINT;
}
exports.isFinal = isFinal;
function toHex(val) {
    if (Buffer.isBuffer(val)) {
        return val.toString('hex');
    }
    if (typeof val === 'string') {
        return val;
    }
    return tools.toHex(val);
}
exports.toHex = toHex;
/**
 * Converts a Opcat address to its public key hash (PKH) in hex format.
 * @param address - The Opcat address to convert.
 * @returns The public key hash as a hex string.
 */
function addrToPkh(address) {
    return tools.toHex(opcat_1.Address.fromString(address).hashBuffer);
}
exports.addrToPkh = addrToPkh;
function pushData(data) {
    const res = [];
    const dLen = data.length;
    if (dLen < 0x4c) {
        const dLenBuff = Buffer.alloc(1);
        dLenBuff.writeUInt8(dLen);
        res.push(dLenBuff);
    }
    else if (dLen <= 0xff) {
        // OP_PUSHDATA1
        res.push(Buffer.from('4c', 'hex'));
        const dLenBuff = Buffer.alloc(1);
        dLenBuff.writeUInt8(dLen);
        res.push(dLenBuff);
    }
    else if (dLen <= 0xffff) {
        // OP_PUSHDATA2
        res.push(Buffer.from('4d', 'hex'));
        const dLenBuff = Buffer.alloc(2);
        dLenBuff.writeUint16LE(dLen);
        res.push(dLenBuff);
    }
    else {
        // OP_PUSHDATA4
        res.push(Buffer.from('4e', 'hex'));
        const dLenBuff = Buffer.alloc(4);
        dLenBuff.writeUint32LE(dLen);
        res.push(dLenBuff);
    }
    res.push(data);
    return Buffer.concat(res);
}
exports.pushData = pushData;
function splitChunks(buffer, chunkSize) {
    const chunks = [];
    for (let i = 0; i < buffer.length; i += chunkSize) {
        chunks.push(buffer.slice(i, i + chunkSize));
    }
    return chunks;
}
exports.splitChunks = splitChunks;
